{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from IPython.display import display\n",
    "from ipywidgets import FloatProgress\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "DIR = r'c://downloads'\n",
    "\n",
    "plt.style.use('ggplot')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PyVCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Phase 3 data from the 1000 Genomes Project (2504 individuals)\n",
    "# From: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/\n",
    "# ALL.chr17.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\n",
    "VCF_FILE_NAME = 'ALL.chr17.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz'\n",
    "vcf_file_path = os.path.join(DIR, VCF_FILE_NAME)\n",
    "\n",
    "with gzip.open(vcf_file_path, 'rt') as f:\n",
    "    print(f.read(100000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "vcf_reader = vcf.Reader(filename = vcf_file_path, compressed = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(vcf_reader.metadata)\n",
    "print('*' * 50)\n",
    "\n",
    "for key, value in vcf_reader.metadata.items():\n",
    "    print('%s: %s' % (key, value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for name, info in vcf_reader.infos.items():\n",
    "    print('%s [%s]: %s' % (name, info.type, info.desc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(vcf_reader.samples))\n",
    "print(vcf_reader.samples[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "record = next(vcf_reader)\n",
    "print('Chromosome %s at %d: %s --> %s (%s)' % (record.CHROM, record.POS, record.REF, record.ALT, record.ID))\n",
    "print(type(record.ALT))\n",
    "print(record.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(record.var_type)\n",
    "print(record.is_snp, record.is_indel)\n",
    "print(record.alleles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(record.samples))\n",
    "print(record.samples[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "call = record.samples[0]\n",
    "print(call)\n",
    "print(call.sample)\n",
    "print(call.gt_alleles)\n",
    "print(call.gt_bases)\n",
    "print(call.is_variant, call.is_het)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(record.genotype('HG00107'))\n",
    "print(record.get_hom_refs()[:3])\n",
    "print(record.get_hets())\n",
    "print(record.get_hom_alts())\n",
    "print(record.get_unknowns())\n",
    "print(record.num_called, record.num_hom_ref, record.num_het, record.num_hom_alt, record.num_unknown)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example: building DNA substitution matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import islice\n",
    "\n",
    "BASES = list('ACGT')\n",
    "\n",
    "dna_sub_matrix = pd.DataFrame(0, index = BASES, columns = BASES)\n",
    "\n",
    "N_RECORDS = 5000\n",
    "progress_bar = FloatProgress(max = N_RECORDS)\n",
    "display(progress_bar)\n",
    "\n",
    "for i, record in enumerate(islice(vcf.Reader(filename = vcf_file_path, compressed = True), N_RECORDS)):\n",
    "\n",
    "    progress_bar.value = i\n",
    "\n",
    "    if record.is_snp:\n",
    "        for alt_allele, frequency in zip(record.ALT, record.INFO['AF']):\n",
    "            dna_sub_matrix.loc[record.REF, alt_allele.sequence] += frequency\n",
    "            \n",
    "dna_sub_matrix = dna_sub_matrix.div(dna_sub_matrix.sum(axis = 1), axis = 0)\n",
    "display(dna_sub_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize = (8, 6))\n",
    "heatmap = ax.pcolor(dna_sub_matrix, cmap = 'cool', vmin = 0)\n",
    "fig.colorbar(heatmap)\n",
    "\n",
    "ax.set_xticks(np.arange(len(BASES)) + 0.5)\n",
    "ax.set_xticklabels(BASES)\n",
    "ax.set_yticks(np.arange(len(BASES)) + 0.5)\n",
    "ax.set_yticklabels(BASES)\n",
    "\n",
    "for i in range(len(BASES)):\n",
    "    for j in range(len(BASES)):\n",
    "        ax.text(i + 0.5, j + 0.5, '%d%%' % (100 * dna_sub_matrix.iloc[i, j]), ha = 'center', va = 'center', color = 'black',\n",
    "                fontsize = 16)\n",
    "\n",
    "_ = ax.set_title('DNA Nucleotide Substitution Matrix')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In-group transitions of Purines (Adenine & Guanine) and Pyrimidines (Cytosine & Thymine) are more common than cross-group transversions."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
