{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import csv\n",
    "import json\n",
    "from collections import defaultdict\n",
    "\n",
    "DIR = r'c://downloads'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing and exploratory analysis of GENCODE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All annotation types: CDS, Selenocysteine, UTR, exon, gene, start_codon, stop_codon, transcript\n",
      "**************************************************\n",
      "All gene types: 3prime_overlapping_ncRNA, IG_C_gene, IG_C_pseudogene, IG_D_gene, IG_J_gene, IG_J_pseudogene, IG_V_gene, IG_V_pseudogene, IG_pseudogene, Mt_rRNA, Mt_tRNA, TEC, TR_C_gene, TR_D_gene, TR_J_gene, TR_J_pseudogene, TR_V_gene, TR_V_pseudogene, antisense, bidirectional_promoter_lncRNA, lincRNA, macro_lncRNA, miRNA, misc_RNA, non_coding, polymorphic_pseudogene, processed_pseudogene, processed_transcript, protein_coding, pseudogene, rRNA, scRNA, sense_intronic, sense_overlapping, snRNA, snoRNA, transcribed_processed_pseudogene, transcribed_unitary_pseudogene, transcribed_unprocessed_pseudogene, translated_processed_pseudogene, unitary_pseudogene, unprocessed_pseudogene, vaultRNA\n",
      "**************************************************\n",
      "Extracted 60880 genes in 37 chromosomes and 1267972 exons in 57781 genes.\n"
     ]
    }
   ],
   "source": [
    "# File taken from:\n",
    "# ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_29/GRCh37_mapping/gencode.v29lift37.annotation.gtf.gz\n",
    "f = gzip.open(os.path.join(DIR, 'gencode.v29lift37.annotation.gtf.gz'), 'rt')\n",
    "csv_reader = csv.reader(f, delimiter = '\\t')\n",
    "\n",
    "for _ in range(5):\n",
    "    next(csv_reader)\n",
    "\n",
    "def parse_extra_fields(raw_extra_fields):\n",
    "\n",
    "    extra_fields = {}\n",
    "\n",
    "    for raw_extra_field in raw_extra_fields[:-1].split(';'):\n",
    "        key, raw_value = raw_extra_field.strip().split(' ')\n",
    "        value = raw_value.strip('\"')\n",
    "        extra_fields[key] = value\n",
    "        \n",
    "    return extra_fields\n",
    "\n",
    "# A dictionary from chromosome name to a list of genes.\n",
    "# Each gene is a tuple of 4 elements: gene_name (str), gene_type (str), start (int), end (int)\n",
    "genes_per_chromosome = defaultdict(list)\n",
    "# A dictionry from gene name to a list of exons.\n",
    "# Each exon is a (start, end) tuple\n",
    "exons_per_gene = defaultdict(list)\n",
    "\n",
    "# We explore the GENCODE database, and are first interested to see what annotation types and gene types are available\n",
    "# within this dataset.\n",
    "all_annotation_types = set()\n",
    "all_gene_types = set()\n",
    "\n",
    "for a_chr, _, a_type, a_start, a_end, _, _, _, raw_extra_fields in csv_reader:\n",
    "\n",
    "    all_annotation_types.add(a_type)\n",
    "    \n",
    "    if a_type == 'gene':\n",
    "        extra_fields = parse_extra_fields(raw_extra_fields)\n",
    "        all_gene_types.add(extra_fields['gene_type'])\n",
    "        genes_per_chromosome[a_chr].append((extra_fields['gene_name'], extra_fields['gene_type'], int(a_start), int(a_end)))\n",
    "    elif a_type == 'exon':\n",
    "        exons_per_gene[extra_fields['gene_name']].append((int(a_start), int(a_end)))\n",
    "        \n",
    "f.close()\n",
    "\n",
    "print('All annotation types: ' + ', '.join(sorted(all_annotation_types)))\n",
    "print('*' * 50)\n",
    "print('All gene types: ' + ', '.join(sorted(all_gene_types)))\n",
    "print('*' * 50)\n",
    "print('Extracted %d genes in %d chromosomes and %d exons in %d genes.' % (sum(map(len, genes_per_chromosome.values())), \\\n",
    "        len(genes_per_chromosome), sum(map(len, exons_per_gene.values())), len(exons_per_gene)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q1A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GL000192.1: 1 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000193.1: 3 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000195.1: 16 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000199.1: 1 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000202.1: 1 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000204.1: 1 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000205.1: 2 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000212.1: 8 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000220.1: 10 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000228.1: 6 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000237.1: 1 genes, of which 0 protein coding and 0 miRNA\n",
      "GL000241.1: 1 genes, of which 0 protein coding and 0 miRNA\n",
      "chr1: 5556 genes, of which 2094 protein coding and 240 miRNA\n",
      "chr10: 2367 genes, of which 753 protein coding and 120 miRNA\n",
      "chr11: 3333 genes, of which 1328 protein coding and 136 miRNA\n",
      "chr12: 3073 genes, of which 1056 protein coding and 122 miRNA\n",
      "chr13: 1375 genes, of which 328 protein coding and 76 miRNA\n",
      "chr14: 2309 genes, of which 630 protein coding and 140 miRNA\n",
      "chr15: 2275 genes, of which 619 protein coding and 118 miRNA\n",
      "chr16: 2593 genes, of which 870 protein coding and 104 miRNA\n",
      "chr17: 3072 genes, of which 1207 protein coding and 112 miRNA\n",
      "chr18: 1233 genes, of which 275 protein coding and 56 miRNA\n",
      "chr19: 3019 genes, of which 1490 protein coding and 158 miRNA\n",
      "chr2: 4160 genes, of which 1275 protein coding and 225 miRNA\n",
      "chr20: 1431 genes, of which 554 protein coding and 86 miRNA\n",
      "chr21: 805 genes, of which 245 protein coding and 31 miRNA\n",
      "chr22: 1364 genes, of which 450 protein coding and 59 miRNA\n",
      "chr3: 3185 genes, of which 1084 protein coding and 191 miRNA\n",
      "chr4: 2634 genes, of which 758 protein coding and 147 miRNA\n",
      "chr5: 2988 genes, of which 899 protein coding and 143 miRNA\n",
      "chr6: 3044 genes, of which 1058 protein coding and 148 miRNA\n",
      "chr7: 2996 genes, of which 926 protein coding and 154 miRNA\n",
      "chr8: 2463 genes, of which 695 protein coding and 138 miRNA\n",
      "chr9: 2456 genes, of which 808 protein coding and 137 miRNA\n",
      "chrM: 37 genes, of which 13 protein coding and 0 miRNA\n",
      "chrX: 2489 genes, of which 853 protein coding and 188 miRNA\n",
      "chrY: 572 genes, of which 68 protein coding and 14 miRNA\n"
     ]
    }
   ],
   "source": [
    "def filter_genes_of_type(genes, gene_type):\n",
    "    return [gene for gene in genes if gene[1] == gene_type]\n",
    "\n",
    "protein_coding_genes_per_chromosome = {}\n",
    "\n",
    "for chromosome, genes in sorted(genes_per_chromosome.items()):\n",
    "    protein_coding_genes = filter_genes_of_type(genes, 'protein_coding')\n",
    "    miRNA_genes = filter_genes_of_type(genes, 'miRNA')\n",
    "    protein_coding_genes_per_chromosome[chromosome] = protein_coding_genes\n",
    "    print('%s: %d genes, of which %d protein coding and %d miRNA' % (chromosome, len(genes), len(protein_coding_genes), \\\n",
    "            len(miRNA_genes)))\n",
    "    \n",
    "f = open(os.path.join(DIR, 'protein_coding_genes_per_chromosome.json'), 'w')\n",
    "json.dump(protein_coding_genes_per_chromosome, f)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q1B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1: avg exon length = 248\n",
      "chr10: avg exon length = 242\n",
      "chr11: avg exon length = 234\n",
      "chr12: avg exon length = 225\n",
      "chr13: avg exon length = 261\n",
      "chr14: avg exon length = 241\n",
      "chr15: avg exon length = 239\n",
      "chr16: avg exon length = 230\n",
      "chr17: avg exon length = 223\n",
      "chr18: avg exon length = 247\n",
      "chr19: avg exon length = 230\n",
      "chr2: avg exon length = 239\n",
      "chr20: avg exon length = 248\n",
      "chr21: avg exon length = 245\n",
      "chr22: avg exon length = 234\n",
      "chr3: avg exon length = 235\n",
      "chr4: avg exon length = 246\n",
      "chr5: avg exon length = 244\n",
      "chr6: avg exon length = 255\n",
      "chr7: avg exon length = 236\n",
      "chr8: avg exon length = 240\n",
      "chr9: avg exon length = 244\n",
      "chrM: avg exon length = 876\n",
      "chrX: avg exon length = 258\n",
      "chrY: avg exon length = 228\n"
     ]
    }
   ],
   "source": [
    "for chromosome, genes in sorted(genes_per_chromosome.items()):\n",
    "\n",
    "    exon_lengths = []\n",
    "\n",
    "    for gene_name, gene_type, _, _ in genes:\n",
    "        if gene_type == 'protein_coding':\n",
    "            for start, end in exons_per_gene[gene_name]:\n",
    "                exon_lengths.append(end - start + 1)\n",
    "                \n",
    "    if len(exon_lengths) > 0:\n",
    "        avg_exon_length = sum(exon_lengths) / len(exon_lengths)\n",
    "        print('%s: avg exon length = %d' % (chromosome, avg_exon_length))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q1C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Longest intergenic region: chrY:28780800-59001390 (length = 30220591 bp)\n"
     ]
    }
   ],
   "source": [
    "longest_intergenic_region_length = 0\n",
    "longest_intergenic_region = None\n",
    "\n",
    "for chromosome, genes in genes_per_chromosome.items():\n",
    "    \n",
    "    # For each chromosome, we extract the coordinates of all genes (regardless of their type), and sort them\n",
    "    # by the start coordinate, so we can then iterate them one after the next with an increasing start coordinate\n",
    "    # (this will be crucial).\n",
    "    gene_coordinates = [(start, end) for _, _, start, end in genes]\n",
    "    gene_coordinates.sort()\n",
    "    \n",
    "    # This variable stores the last cooridnate along the chromosome we have processed so far. For example, if we have\n",
    "    # last_processed_coordinate = 7,123,456, then it means we have already processed all the intergenic regions until coordinate\n",
    "    # 7,123,456 on the chromosome and there is a gene ending in that coordinate. \n",
    "    last_processed_coordinate = 0\n",
    "    \n",
    "    for start, end in gene_coordinates:\n",
    "        \n",
    "        # The last genes we processed ended in 'last_processed_coordinate', and the current gene starts in 'start' coordinate.\n",
    "        # Therefore, this expression calculates the length of the intergenic region we have just encountered.\n",
    "        # Note that it might be that start < last_processed_coordinate, if we have just encountered a gene overlapping with the\n",
    "        # last gene. In that case, 'intergenic_length' will become a negative number, and the next if statement won't run.\n",
    "        # Therefore, it is not an issue.\n",
    "        # Remember we are dealing with 1-based indices. \n",
    "        intergenic_length = start - last_processed_coordinate - 1\n",
    "\n",
    "        if intergenic_length > longest_intergenic_region_length:\n",
    "            # We have just broken the record and found the biggest intergenic regiob thus far.\n",
    "            longest_intergenic_region_length = intergenic_length\n",
    "            longest_intergenic_region = (chromosome, last_processed_coordinate + 1, start - 1)\n",
    "                \n",
    "        # Now we have processed all the genetic loci up entil the 'end' coordinate of the current gene.\n",
    "        # In rare instances (when we have overlapping genes), it might be that end < last_gene_coordinate (i.e. the\n",
    "        # current gene overlaps with the last processed gene but ends before it). That's the reason why we take\n",
    "        # the maximum between the two. \n",
    "        last_processed_coordinate = max(last_processed_coordinate, end)\n",
    "        \n",
    "print('Longest intergenic region: %s:%d-%d (length = %d bp)' % (longest_intergenic_region + \\\n",
    "        (longest_intergenic_region_length,)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "According to this analysis, the longest intergenic region is a ~30Mbp region on chromosome Y.\n",
    "\n",
    "Of course, the very notion of \"intergenic region\" is very ill-defined. Crucially, it depends on (i) what we define as a gene, and (ii) what gene annotations we are using. The notion of \"longest intergenic region\" is particularly sensitive, because it is sufficient to have one additional gene annotation (in the middle of that region) to cut its length in half, deeming it no longer the longest intergenic region.\n",
    "\n",
    "In our case, we relied on GENCODE (release 29, mapped to version GRCh37 of the human reference genome); we considered all the genomic elements describes as \"genes\". It is very likely that if we used different annotations (even just a different version of GENCODE) we would get a totally different genomic region, due to the considerations mentioned above."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
