{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "f = open(r'c://downloads/herpesvirus_genome.json', 'r')\n",
    "data = json.load(f)\n",
    "f.close()\n",
    "print(type(data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# It's safer to check how big something is before printing it.\n",
    "print(len(data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, value in data.items():\n",
    "    print(key, type(value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, value in data.items():\n",
    "    if key != 'coding_regions':\n",
    "        print('%s: %s' % (key, value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_regions = data['coding_regions']\n",
    "print(type(coding_regions))\n",
    "print(len(coding_regions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_region = coding_regions[0]\n",
    "print(type(coding_region))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(coding_region))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, value in coding_region.items():\n",
    "    print(key, type(value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, value in coding_region.items():\n",
    "    if key != 'intervals':\n",
    "        print('%s: %s' % (key, value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(coding_region['intervals']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interval, = coding_region['intervals']\n",
    "print(type(interval))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(interval))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, value in interval.items():\n",
    "    print(key, type(value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(interval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(coding_region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "products = []\n",
    "\n",
    "for coding_region in coding_regions:\n",
    "    products.append(coding_region['product'])\n",
    "    \n",
    "print(products)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lengths_per_group = {'envelope': [], 'membrane': [], 'capsid': []}\n",
    "all_lengths = []\n",
    "\n",
    "for coding_region in coding_regions:\n",
    "    \n",
    "    product_name = coding_region['product'].lower()\n",
    "    length = len(coding_region['translation'])\n",
    "    \n",
    "    for group_name, group_lengths in lengths_per_group.items():\n",
    "        if group_name in product_name:\n",
    "            group_lengths.append(length)\n",
    "    \n",
    "    all_lengths.append(length)\n",
    "    \n",
    "lengths_per_group['all'] = all_lengths\n",
    "\n",
    "for group_name, group_lengths in sorted(lengths_per_group.items()):\n",
    "    avg = sum(group_lengths) / len(group_lengths)\n",
    "    print('%s: # = %d, avg. = %.2f aa' % (group_name, len(group_lengths), avg))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(r'c://downloads/protein_lengths_per_group.json', 'w')\n",
    "json.dump(lengths_per_group, f)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_json = json.dumps(lengths_per_group)\n",
    "print(type(raw_json))\n",
    "print(raw_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = json.loads(raw_json)\n",
    "print(type(data))\n",
    "print(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading human gene annotations from gencode.v29lift37.annotation.gtf.gz at:\n",
    "# https://www.gencodegenes.org/human/release_29lift37.html\n",
    "\n",
    "import gzip\n",
    "\n",
    "f = gzip.open(r'c://downloads/gencode.v29lift37.annotation.gtf.gz', 'rt')\n",
    "print(f.read(1000))\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# When using gzip with Python 3, you need to explictly ask for text (t) mode\n",
    "f = gzip.open(r'c://downloads/gencode.v29lift37.annotation.gtf.gz', 'r')\n",
    "print(f.read(1000))\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "f = gzip.open(r'c://downloads/gencode.v29lift37.annotation.gtf.gz', 'rt')\n",
    "\n",
    "# Default delimiter is comma (,)\n",
    "csv_reader = csv.reader(f, delimiter = '\\t')\n",
    "\n",
    "# Skip 5 first header lines \n",
    "for _ in range(5):\n",
    "    next(csv_reader)\n",
    "    \n",
    "annotations = []\n",
    "\n",
    "for _ in range(100):\n",
    "    annotations.append(next(csv_reader))\n",
    "\n",
    "f.close()\n",
    "\n",
    "print(len(annotations))\n",
    "print(annotations[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If want to go over all lines, just iterate over the csv reader with a for loop (no need to use the 'next' function)\n",
    "# (this will take too long for this demo)\n",
    "\n",
    "for line in csv_reader:\n",
    "    # Do something...\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = []\n",
    "\n",
    "def parse_extra_fields(raw_extra_fields):\n",
    "\n",
    "    extra_fields = {}\n",
    "\n",
    "    for raw_extra_field in raw_extra_fields[:-1].split(';'):\n",
    "        key, raw_value = raw_extra_field.strip().split(' ')\n",
    "        value = raw_value.strip('\"')\n",
    "        extra_fields[key] = value\n",
    "        \n",
    "    return extra_fields\n",
    "\n",
    "for a_chr, _, a_type, a_start, a_end, _, _, _, raw_extra_fields in annotations:\n",
    "    if a_type == 'gene':\n",
    "        extra_fields = parse_extra_fields(raw_extra_fields)\n",
    "        genes.append([extra_fields['gene_name'], extra_fields['gene_type'], a_chr, int(a_start), int(a_end)])\n",
    "\n",
    "print(len(genes))\n",
    "print(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(r'c://downloads/genes.csv', 'w', newline = '')\n",
    "csv_writer = csv.writer(f)\n",
    "csv_writer.writerows(genes)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notes:\n",
    "* `newline = ''` is for avoiding blank lines between each two content lines\n",
    "* `writerows` expects a list of lists of strings (or objects to convert to strings)\n",
    "* You can write one row at a time using `writerow`, which expects a list of strings"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
