{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3d453ed1",
   "metadata": {},
   "source": [
    "## Lesson 07 Homework Exercises"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46d3ef1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "from openpyxl import load_workbook\n",
    "# DIR = '/Users/fernandojuarez/Documents/GitHub/Quantitative_Biological_Research_Python_course/'\n",
    "DIR = '/Users/fernandojuarez/Documents/GitHub/Quantitative_Biological_Research_Python_course/QBRP/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaa60335",
   "metadata": {},
   "source": [
    "# Exercise 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61e39634",
   "metadata": {},
   "source": [
    "## Part A\n",
    "Download from Pfam’s FTP site all the occurrences of Pfam domains/HMM profiles in the entire human\n",
    "proteome. Use version 31.0 (not the newer ones). It should be a compressed .tsv file with ~100k rows\n",
    "and 14 columns. Each row in this file indicates an occurrence of a single domain/HMM profile in a single\n",
    "protein (each profile can occur multiple times in different proteins, and even in the same protein, and\n",
    "each protein can have multiple profiles).\n",
    "\n",
    "Among the fields in this file are:\n",
    "\n",
    "● seq id: the UniProt ID of the protein where the profile occurs.\n",
    "\n",
    "● alignment start & alignment end: 1-based coordinates indicating the position of the profile within the protein.\n",
    "\n",
    "● hmm acc & hmm name: The accession ID and the name of the profile occurring in the protein.\n",
    "\n",
    "● type: the type of the occurring HMM profile (most commonly Domain).\n",
    "\n",
    "How many occurrences are there of each type?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00b17209",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e448458e",
   "metadata": {},
   "source": [
    "## Part B\n",
    "Filter only the records of type Domain. What is the average length of Pfam domains? What are the 5\n",
    "longest and 5 shortest domains (on average)?\n",
    "\n",
    "Note that the field hmm_length indicates the theoretical length of the Pfam profile, not the actual\n",
    "length of its occurrence within a given protein."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "383a4efb",
   "metadata": {},
   "source": [
    "## Part C\n",
    "Compare the lengths of the G-alpha and tyrosine kinase domains. Is there a large difference in their\n",
    "average lengths?\n",
    "\n",
    "Is this difference significant? Consider using Scipy’s t-test [https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html]\n",
    "\n",
    "(see this video [https://www.youtube.com/watch?v=0Pd3dc1GcHc] for a refresher on t-tests). Other\n",
    "appropriate statistical tests will also be accepted."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd4f1bd8",
   "metadata": {},
   "source": [
    "## Part D\n",
    "Download the required protein sequences from UniProt and add a new column to the table for the\n",
    "entire protein sequences of the records. Add another column for the sequence of the HMM profile\n",
    "within the protein (i.e. the latter is a subsequence of the former)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1049fdd",
   "metadata": {},
   "source": [
    "## Part E\n",
    "What is the biological role of the PDZ domain?\n",
    "\n",
    "Save the sequences of all the occurrences of the PDZ domain as a FASTA file. Submit it to an online\n",
    "https://www.ebi.ac.uk/Tools/msa/clustalo/) and create a logo\n",
    "Multiple Sequence Alignment tool (e.g. from it (e.g. using http://www.cbs.dtu.dk/biotools/Seq2Logo-2.1/).\n",
    "\n",
    "Compare your logo to Pfam’s HMM logo for this domain. Identify similarities and differences between\n",
    "the two logos, and suggest plausible explanations for the differences."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf8d85ca",
   "metadata": {},
   "source": [
    "## Part F\n",
    "There’s one protein with 9 occurrences of the PDZ domain. Find it.\n",
    "\n",
    "What is known about the function of this protein?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a34602e",
   "metadata": {},
   "source": [
    "## Part G [Optional, recommended]\n",
    "\n",
    "What percentage of the exon junctions in the coding region of the transcript of the protein you have\n",
    "found occur within a PDZ domain? What percentage would you expect to see at random?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04bbfd17",
   "metadata": {},
   "source": [
    "## Part H [Optional, recommended]\n",
    "At which locations inside the PDZ domain do exon junctions occur (across the entire human proteome)?\n",
    "How are they distributed compared to random uniform distribution?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbb57785",
   "metadata": {},
   "source": [
    "## Part I [Optional, recommended]\n",
    "Do PDZ occurrences within the same protein more similar to each other than occurrences in different\n",
    "proteins?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce4a9520",
   "metadata": {},
   "source": [
    "# Bonus Question 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce575fa8",
   "metadata": {},
   "source": [
    "What is “time series” data and why is it important in biological and medical research?\n",
    "\n",
    "Demonstrate Panda’s time-series capabilities with an example biological/medical dataset (e.g. you may\n",
    "use the BioTIME dataset [http://biotime.st-andrews.ac.uk/home.php])."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
