{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q1\n",
    "\n",
    "a) **A DNA sequence** - str (string), with each letter in the string representing a single nucleotide ('A', 'C', 'G' or 'T'). Another option, somewhat less convenient, is to use a list of single-letter strings (nucleotides).\n",
    "\n",
    "b) **Names/symbols of genes associated with a disease** - A set of strings, with each element in the set being the name/symbol of a specific gene. Since generally there is no relevance for the order of the genes (and they are unique, namely this is meaningless to include a gene twice in such a list), it makes more sense to use a set than a list.\n",
    "\n",
    "c) **RNA sequences of all the transcripts expressed in a tissue** - If there is no relevance to the order of the transcripts, it would be the most convenient to use a dictionary of the form `{'name_of_transcript1': 'seq1', 'name_of_transcript2': 'seq2', ...}` (i.e. the keys and values of the dictionary are of type str). If the order of transcripts is relevant, then can use a list of tuples, i.e. `[('name_of_transcript1', 'seq1'), ('name_of_transcript2', 'seq2'), ...]`. If we just care about the sequences themselves (and not about the names of the transcripts), then we can use a list of the form `['seq1', 'seq2', ...]`.\n",
    "\n",
    "d) **Frequencies of amino acids in a given protein sequence** - A dictionary of the form `{'aa1': freq1, 'aa2': freq2, ...}`, for example `{'A': 0.2, 'K': 0.1, 'L': 0.4, 'P': 0.15, 'G': 0.15}`. A dictionary is the most convenient data structure for representing a mapping, like  in this case (mapping the frequency of each amino acid), given that the keys are unique and their order is irrelevant.\n",
    "\n",
    "e) **Unique amino-acids appearing in a short peptide** - A set of strings, each a one letter representing the unique amino acid. For example, `{'A', 'K', 'L', 'P', 'G'}`. A set is the right data structure to use for representing unique elements (whose order is irrelevant)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q2A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After 1 days: strain A = 20000 (98.5%), strain B = 300 (1.5%) [Total = 20300]\n",
      "After 2 days: strain A = 40000 (97.8%), strain B = 900 (2.2%) [Total = 40900]\n",
      "After 3 days: strain A = 80000 (96.7%), strain B = 2700 (3.3%) [Total = 82700]\n",
      "After 4 days: strain A = 160000 (95.2%), strain B = 8100 (4.8%) [Total = 168100]\n",
      "After 5 days: strain A = 320000 (92.9%), strain B = 24300 (7.1%) [Total = 344300]\n",
      "After 6 days: strain A = 640000 (89.8%), strain B = 72900 (10.2%) [Total = 712900]\n",
      "After 7 days: strain A = 1280000 (85.4%), strain B = 218700 (14.6%) [Total = 1498700]\n",
      "After 8 days: strain A = 2560000 (79.6%), strain B = 656100 (20.4%) [Total = 3216100]\n",
      "After 9 days: strain A = 5120000 (72.2%), strain B = 1968300 (27.8%) [Total = 7088300]\n",
      "After 10 days: strain A = 10240000 (63.4%), strain B = 5904900 (36.6%) [Total = 16144900]\n",
      "After 11 days: strain A = 20480000 (53.6%), strain B = 17714700 (46.4%) [Total = 38194700]\n",
      "After 12 days: strain A = 40960000 (43.5%), strain B = 53144100 (56.5%) [Total = 94104100]\n",
      "After 13 days: strain A = 81920000 (33.9%), strain B = 159432300 (66.1%) [Total = 241352300]\n",
      "After 14 days: strain A = 122880000 (30.0%), strain B = 286978140 (70.0%) [Total = 409858140]\n",
      "After 15 days: strain A = 184320000 (26.3%), strain B = 516560652 (73.7%) [Total = 700880652]\n",
      "After 16 days: strain A = 276480000 (22.9%), strain B = 929809173 (77.1%) [Total = 1206289173]\n",
      "After 17 days: strain A = 414720000 (19.9%), strain B = 1673656512 (80.1%) [Total = 2088376512]\n",
      "After 18 days: strain A = 622080000 (17.1%), strain B = 3012581722 (82.9%) [Total = 3634661722]\n",
      "After 19 days: strain A = 933120000 (14.7%), strain B = 5422647100 (85.3%) [Total = 6355767100]\n",
      "After 20 days: strain A = 1399680000 (12.5%), strain B = 9760764780 (87.5%) [Total = 11160444780]\n"
     ]
    }
   ],
   "source": [
    "strain_a = 1e04\n",
    "strain_b = 1e02\n",
    "day = 0\n",
    "\n",
    "while strain_a + strain_b < 1e10:\n",
    "    \n",
    "    if strain_a + strain_b < 1e08:\n",
    "        strain_a *= 2\n",
    "        strain_b *= 3\n",
    "    else:\n",
    "        strain_a *= 1.5\n",
    "        strain_b *= 1.8\n",
    "    \n",
    "    day += 1\n",
    "    total = strain_a + strain_b\n",
    "    print('After %d days: strain A = %d (%.1f%%), strain B = %d (%.1f%%) [Total = %d]' % (day, strain_a, \\\n",
    "            100 * strain_a / total, strain_b, 100 * strain_b / total, total))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q2B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After 1 days:\n",
      "\tUnimmune A: 20000 bacteria (98.5%)\n",
      "\tImmune A: 0 bacteria (0.0%)\n",
      "\tUnimmune B: 300 bacteria (1.5%)\n",
      "\tImmune B: 0 bacteria (0.0%)\n",
      "\tTotal: 20300 bacteria\n",
      "After 2 days:\n",
      "\tUnimmune A: 40000 bacteria (97.8%)\n",
      "\tImmune A: 0 bacteria (0.0%)\n",
      "\tUnimmune B: 900 bacteria (2.2%)\n",
      "\tImmune B: 0 bacteria (0.0%)\n",
      "\tTotal: 40900 bacteria\n",
      "After 3 days:\n",
      "\tUnimmune A: 80000 bacteria (96.7%)\n",
      "\tImmune A: 0 bacteria (0.0%)\n",
      "\tUnimmune B: 2700 bacteria (3.3%)\n",
      "\tImmune B: 0 bacteria (0.0%)\n",
      "\tTotal: 82700 bacteria\n",
      "After 4 days:\n",
      "\tUnimmune A: 160000 bacteria (95.2%)\n",
      "\tImmune A: 0 bacteria (0.0%)\n",
      "\tUnimmune B: 8100 bacteria (4.8%)\n",
      "\tImmune B: 0 bacteria (0.0%)\n",
      "\tTotal: 168100 bacteria\n",
      "After 5 days:\n",
      "\tUnimmune A: 320000 bacteria (92.9%)\n",
      "\tImmune A: 0 bacteria (0.0%)\n",
      "\tUnimmune B: 24300 bacteria (7.1%)\n",
      "\tImmune B: 0 bacteria (0.0%)\n",
      "\tTotal: 344300 bacteria\n",
      "After 6 days:\n",
      "\tUnimmune A: 640000 bacteria (89.8%)\n",
      "\tImmune A: 0 bacteria (0.0%)\n",
      "\tUnimmune B: 72900 bacteria (10.2%)\n",
      "\tImmune B: 0 bacteria (0.0%)\n",
      "\tTotal: 712900 bacteria\n",
      "After 7 days:\n",
      "\tUnimmune A: 1280000 bacteria (85.4%)\n",
      "\tImmune A: 0 bacteria (0.0%)\n",
      "\tUnimmune B: 218700 bacteria (14.6%)\n",
      "\tImmune B: 0 bacteria (0.0%)\n",
      "\tTotal: 1498700 bacteria\n",
      "After 8 days [active phage]:\n",
      "\tUnimmune A: 1382400 bacteria (71.6%)\n",
      "\tImmune A: 153600 bacteria (8.0%)\n",
      "\tUnimmune B: 389723 bacteria (20.2%)\n",
      "\tImmune B: 3936 bacteria (0.2%)\n",
      "\tTotal: 1929660 bacteria\n",
      "After 9 days [active phage]:\n",
      "\tUnimmune A: 1492992 bacteria (55.7%)\n",
      "\tImmune A: 473088 bacteria (17.7%)\n",
      "\tUnimmune B: 694487 bacteria (25.9%)\n",
      "\tImmune B: 18824 bacteria (0.7%)\n",
      "\tTotal: 2679391 bacteria\n",
      "After 10 days [active phage]:\n",
      "\tUnimmune A: 1612431 bacteria (39.9%)\n",
      "\tImmune A: 1125335 bacteria (27.8%)\n",
      "\tUnimmune B: 1237576 bacteria (30.6%)\n",
      "\tImmune B: 68975 bacteria (1.7%)\n",
      "\tTotal: 4044317 bacteria\n",
      "After 11 days [active phage]:\n",
      "\tUnimmune A: 1741425 bacteria (26.3%)\n",
      "\tImmune A: 2444161 bacteria (36.9%)\n",
      "\tUnimmune B: 2205360 bacteria (33.3%)\n",
      "\tImmune B: 229202 bacteria (3.5%)\n",
      "\tTotal: 6620150 bacteria\n",
      "After 12 days [active phage]:\n",
      "\tUnimmune A: 1880739 bacteria (16.2%)\n",
      "\tImmune A: 5097294 bacteria (43.8%)\n",
      "\tUnimmune B: 3929952 bacteria (33.8%)\n",
      "\tImmune B: 727302 bacteria (6.3%)\n",
      "\tTotal: 11635289 bacteria\n",
      "After 13 days [active phage]:\n",
      "\tUnimmune A: 2031199 bacteria (9.4%)\n",
      "\tImmune A: 10420278 bacteria (48.0%)\n",
      "\tUnimmune B: 7003175 bacteria (32.3%)\n",
      "\tImmune B: 2252647 bacteria (10.4%)\n",
      "\tTotal: 21707299 bacteria\n",
      "After 14 days [active phage]:\n",
      "\tUnimmune A: 2193695 bacteria (5.1%)\n",
      "\tImmune A: 21084300 bacteria (49.4%)\n",
      "\tUnimmune B: 12479657 bacteria (29.3%)\n",
      "\tImmune B: 6883998 bacteria (16.1%)\n",
      "\tTotal: 42641652 bacteria\n",
      "After 15 days [active phage]:\n",
      "\tUnimmune A: 2369190 bacteria (2.7%)\n",
      "\tImmune A: 42431844 bacteria (48.3%)\n",
      "\tUnimmune B: 22238750 bacteria (25.3%)\n",
      "\tImmune B: 20876629 bacteria (23.7%)\n",
      "\tTotal: 87916415 bacteria\n",
      "After 16 days [active phage]:\n",
      "\tUnimmune A: 2558725 bacteria (1.3%)\n",
      "\tImmune A: 85147992 bacteria (44.7%)\n",
      "\tUnimmune B: 39629453 bacteria (20.8%)\n",
      "\tImmune B: 63030186 bacteria (33.1%)\n",
      "\tTotal: 190366358 bacteria\n",
      "After 17 days [active phage]:\n",
      "\tUnimmune A: 2072567 bacteria (0.7%)\n",
      "\tImmune A: 127952273 bacteria (44.7%)\n",
      "\tUnimmune B: 42371811 bacteria (14.8%)\n",
      "\tImmune B: 113882334 bacteria (39.8%)\n",
      "\tTotal: 286278987 bacteria\n",
      "After 18 days [active phage]:\n",
      "\tUnimmune A: 1678780 bacteria (0.4%)\n",
      "\tImmune A: 192114941 bacteria (43.2%)\n",
      "\tUnimmune B: 45303940 bacteria (10.2%)\n",
      "\tImmune B: 205445817 bacteria (46.2%)\n",
      "\tTotal: 444543480 bacteria\n",
      "After 19 days [active phage]:\n",
      "\tUnimmune A: 1359811 bacteria (0.2%)\n",
      "\tImmune A: 288323502 bacteria (40.7%)\n",
      "\tUnimmune B: 48438973 bacteria (6.8%)\n",
      "\tImmune B: 370291754 bacteria (52.3%)\n",
      "\tTotal: 708414042 bacteria\n",
      "After 20 days [active phage]:\n",
      "\tUnimmune A: 1101447 bacteria (0.1%)\n",
      "\tImmune A: 432607637 bacteria (37.5%)\n",
      "\tUnimmune B: 51790950 bacteria (4.5%)\n",
      "\tImmune B: 667048298 bacteria (57.9%)\n",
      "\tTotal: 1152548334 bacteria\n",
      "After 21 days [active phage]:\n",
      "\tUnimmune A: 892172 bacteria (0.0%)\n",
      "\tImmune A: 649010586 bacteria (34.0%)\n",
      "\tUnimmune B: 55374884 bacteria (2.9%)\n",
      "\tImmune B: 1201246280 bacteria (63.0%)\n",
      "\tTotal: 1906523923 bacteria\n",
      "After 22 days [active phage]:\n",
      "\tUnimmune A: 722659 bacteria (0.0%)\n",
      "\tImmune A: 973596174 bacteria (30.5%)\n",
      "\tUnimmune B: 59206826 bacteria (1.9%)\n",
      "\tImmune B: 2162841352 bacteria (67.7%)\n",
      "\tTotal: 3196367013 bacteria\n",
      "After 23 days [active phage]:\n",
      "\tUnimmune A: 585354 bacteria (0.0%)\n",
      "\tImmune A: 1460459301 bacteria (27.0%)\n",
      "\tUnimmune B: 63303938 bacteria (1.2%)\n",
      "\tImmune B: 3893753869 bacteria (71.9%)\n",
      "\tTotal: 5418102463 bacteria\n",
      "After 24 days [active phage]:\n",
      "\tUnimmune A: 474137 bacteria (0.0%)\n",
      "\tImmune A: 2190741634 bacteria (23.6%)\n",
      "\tUnimmune B: 67684571 bacteria (0.7%)\n",
      "\tImmune B: 7009440646 bacteria (75.6%)\n",
      "\tTotal: 9268340989 bacteria\n",
      "After 25 days [active phage]:\n",
      "\tUnimmune A: 384051 bacteria (0.0%)\n",
      "\tImmune A: 3286155123 bacteria (20.6%)\n",
      "\tUnimmune B: 72368343 bacteria (0.5%)\n",
      "\tImmune B: 12617724157 bacteria (79.0%)\n",
      "\tTotal: 15976631676 bacteria\n"
     ]
    }
   ],
   "source": [
    "# The fraction of bacteria that develop immunity, for each of the two strains.\n",
    "IMMUNITY_DEVELOPMENT_RATES = {\n",
    "    'A': 0.1,\n",
    "    'B': 0.01,\n",
    "}\n",
    "\n",
    "# Number of bacteria at the beginning of the experiment (for each of the 4 populations)\n",
    "populations = {\n",
    "    # (strain, is_immune): population_size\n",
    "    ('A', False): 1e04,\n",
    "    ('A', True): 0,\n",
    "    ('B', False): 1e02,\n",
    "    ('B', True): 0,\n",
    "}\n",
    "\n",
    "day = 0\n",
    "is_phage_active = False\n",
    "\n",
    "# sum(population.values()) provides the total number of bacteria (of all 4 populations)\n",
    "while sum(populations.values()) < 1e10:\n",
    "\n",
    "    # Determine if the phage should turn active (note that once it activates, it cannot deactivate)\n",
    "    if sum(populations.values()) >= 1e06:\n",
    "        is_phage_active = True\n",
    "\n",
    "    # Determine the base growth rates, according to the total population size\n",
    "    if sum(populations.values()) < 1e08:\n",
    "        growth_rate = {'A': 2, 'B': 3}\n",
    "    else:\n",
    "        growth_rate = {'A': 1.5, 'B': 1.8}\n",
    "\n",
    "    # Update the size of each of the 4 populations\n",
    "    for population_key in populations.keys():\n",
    "    \n",
    "        strain, is_immune = population_key\n",
    "        \n",
    "        # Multiply the population size by the appropriate growth rate of the strain.\n",
    "        populations[population_key] *= growth_rate[strain]\n",
    "\n",
    "        # If the phage is active and the population is not immune, then 40% of the cells will die. Note that it doesn't\n",
    "        # matter if we first multiply by the growth rate and then multiply by the phage decay factor or do it in a reversed\n",
    "        # order (because multiplication is symmetric).\n",
    "        if is_phage_active and not is_immune:\n",
    "            populations[population_key] *= 0.6\n",
    "                \n",
    "    if is_phage_active:\n",
    "        for strain in ['A', 'B']:\n",
    "            immunity_rate = IMMUNITY_DEVELOPMENT_RATES[strain]\n",
    "            n_bacteria_developing_immunity = immunity_rate * populations[(strain, False)]\n",
    "            populations[(strain, True)] += n_bacteria_developing_immunity\n",
    "            # Note that the number of bacteria that develop immunity should be deducted from the unimmune population!\n",
    "            populations[(strain, False)] -= n_bacteria_developing_immunity\n",
    "    \n",
    "    # Print the report for the end of the day.\n",
    "    \n",
    "    day += 1\n",
    "    \n",
    "    if is_phage_active:\n",
    "        print('After %d days [active phage]:' % day)\n",
    "    else:\n",
    "        print('After %d days:' % day)\n",
    "    \n",
    "    for strain in ['A', 'B']:\n",
    "        for is_immune in [False, True]:\n",
    "            \n",
    "            if is_immune:\n",
    "                immune_label = 'Immune'\n",
    "            else:\n",
    "                immune_label = 'Unimmune'\n",
    "                \n",
    "            population_size = populations[(strain, is_immune)]\n",
    "            population_percentage = 100 * population_size / sum(populations.values())\n",
    "            print('\\t' + '%s %s: %d bacteria (%.1f%%)' % (immune_label, strain, population_size, population_percentage))\n",
    "            \n",
    "    print('\\t' + 'Total: %d bacteria' % sum(populations.values()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By the end of the experiment, the phage-immune populations appear to have taken over the colony. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The reverse complement of the given DNA sequence is: ATGCCCCAGACTGTGAGTTGCAACAGATGCTGTTGCGCCTCAATAGCCCTCAGCAAATTGTTCTGCTGCTGCACTATACCAGACAATAATTGTCTGGCCTGTACCGTCAGCGTCATTGAGGCTGCGCCCATAGTGCTTCCTGCTGCTCCCAAGAACCCAAGGAACAAAGCTCCTATTCCCACTGCTCTTTTTTCTCTCTGCACCACTCTTCTCTTTGCCTTGGTGGGTGCTACTCCTAATGGTTCAATTTTTACTACTTTATATTTATATAATTCACTTCTCCAATTGTCCCTCATATCTCCTCCTCCAGGTCTGAAGATCTCGGACTCATTGTTGCTATTACCACCATCTCTTGTTAATAGCAGCCCTGTAATATTTGATGAACATCTAATTTGTCCACTGATGGGAGGGGCATACATTGCTTTTCCTACTTTCTGCCACATGTTTATAATTTGTTTTATTCTGCATGGGAGGGTGATTGTGTCACTTCCTTCAGTGTTATTTGACCCTTCAGTACTCCAAGTACTATTAAACCAAGTACTATTAAACAGTTGTGTTGAATTACAGTAG\n",
      "The resulted protein sequence: MPQTVSCNRCCCASIALSKLFCCCTIPDNNCLACTVSVIEAAPIVLPAAPKNPRNKAPIPTALFSLCTTLLFALVGATPNGSIFTTLYLYNSLLQLSLISPPPGLKISDSLLLLPPSLVNSSPVIFDEHLICPLMGGAYIAFPTFCHMFIICFILHGRVIVSLPSVLFDPSVLQVLLNQVLLNSCVELQ\n"
     ]
    }
   ],
   "source": [
    "DNA_seq = 'CTACTGTAATTCAACACAACTGTTTAATAGTACTTGGTTTAATAGTACTTGGAGTACTGAAGGGTCAAATAACACTGAAGGAAGTGACACAATCACCCTCCCATGCAGAAT' + \\\n",
    "          'AAAACAAATTATAAACATGTGGCAGAAAGTAGGAAAAGCAATGTATGCCCCTCCCATCAGTGGACAAATTAGATGTTCATCAAATATTACAGGGCTGCTATTAACAAGAGA' + \\\n",
    "          'TGGTGGTAATAGCAACAATGAGTCCGAGATCTTCAGACCTGGAGGAGGAGATATGAGGGACAATTGGAGAAGTGAATTATATAAATATAAAGTAGTAAAAATTGAACCATT' + \\\n",
    "          'AGGAGTAGCACCCACCAAGGCAAAGAGAAGAGTGGTGCAGAGAGAAAAAAGAGCAGTGGGAATAGGAGCTTTGTTCCTTGGGTTCTTGGGAGCAGCAGGAAGCACTATGGG' + \\\n",
    "          'CGCAGCCTCAATGACGCTGACGGTACAGGCCAGACAATTATTGTCTGGTATAGTGCAGCAGCAGAACAATTTGCTGAGGGCTATTGAGGCGCAACAGCATCTGTTGCAACT' + \\\n",
    "          'CACAGTCTGGGGCAT'\n",
    "\n",
    "# Calcluate the reverse complement of the sequence.\n",
    "\n",
    "COMPLEMENT_NTS = {\n",
    "    'A': 'T',\n",
    "    'C': 'G',\n",
    "    'G': 'C',\n",
    "    'T': 'A',\n",
    "}\n",
    "          \n",
    "reverse_complement_DNA_seq = []\n",
    "\n",
    "# Go over the given DNA seq in reversed order, and for each nucleotide, append its complement nucleotide to the resulting \n",
    "# DNA sequence (to obtain the reverse complement)\n",
    "for nt in DNA_seq[::-1]:\n",
    "    reverse_complement_DNA_seq.append(COMPLEMENT_NTS[nt])\n",
    "    \n",
    "# Convert the result from a list of letters into one string\n",
    "reverse_complement_DNA_seq = ''.join(reverse_complement_DNA_seq)\n",
    "\n",
    "print('The reverse complement of the given DNA sequence is: %s' % reverse_complement_DNA_seq)\n",
    "\n",
    "\n",
    "# Translate the resulted sequence into protein.\n",
    "\n",
    "CODON_TABLE = {\n",
    "    'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "    'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
    "    'UAU': 'Y', 'UAC': 'Y', 'UAA': '*', 'UAG': '*',\n",
    "    'UGU': 'C', 'UGC': 'C', 'UGA': '*', 'UGG': 'W',\n",
    "    'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "    'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
    "    'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
    "    'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'GAU': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G',\n",
    "}\n",
    "\n",
    "reverse_complement_RNA_seq = reverse_complement_DNA_seq.replace('T', 'U')\n",
    "\n",
    "protein_seq = []\n",
    "\n",
    "for i in range(0, len(reverse_complement_RNA_seq), 3):\n",
    "\n",
    "    codon = reverse_complement_RNA_seq[i:(i + 3)]\n",
    "    aa = CODON_TABLE[codon]\n",
    "\n",
    "    if aa == '*':\n",
    "        break\n",
    "    else:\n",
    "        protein_seq.append(aa)\n",
    "\n",
    "protein_seq = ''.join(protein_seq)\n",
    "print('The resulted protein sequence: %s' % protein_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q3B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Codon frequencies: {'AUG': 0.015789473684210527, 'CCC': 0.021052631578947368, 'CAG': 0.010526315789473684, 'ACU': 0.042105263157894736, 'GUG': 0.031578947368421054, 'AGU': 0.010526315789473684, 'UGC': 0.042105263157894736, 'AAC': 0.02631578947368421, 'AGA': 0.005263157894736842, 'UGU': 0.031578947368421054, 'GCC': 0.021052631578947368, 'UCA': 0.03684210526315789, 'AUA': 0.031578947368421054, 'CUC': 0.031578947368421054, 'AGC': 0.021052631578947368, 'AAA': 0.010526315789473684, 'UUG': 0.02631578947368421, 'UUC': 0.010526315789473684, 'CCA': 0.031578947368421054, 'GAC': 0.015789473684210527, 'AAU': 0.02631578947368421, 'CUG': 0.021052631578947368, 'ACC': 0.010526315789473684, 'GUC': 0.010526315789473684, 'AUU': 0.042105263157894736, 'GAG': 0.005263157894736842, 'GCU': 0.03684210526315789, 'GCG': 0.005263157894736842, 'CUU': 0.031578947368421054, 'CCU': 0.04736842105263158, 'AAG': 0.010526315789473684, 'AGG': 0.010526315789473684, 'UUU': 0.042105263157894736, 'UCU': 0.015789473684210527, 'GGU': 0.015789473684210527, 'UUA': 0.03684210526315789, 'UAU': 0.010526315789473684, 'CAA': 0.015789473684210527, 'UCC': 0.005263157894736842, 'AUC': 0.005263157894736842, 'UCG': 0.005263157894736842, 'CUA': 0.021052631578947368, 'GUU': 0.010526315789473684, 'GUA': 0.021052631578947368, 'GAU': 0.005263157894736842, 'GAA': 0.010526315789473684, 'CAU': 0.010526315789473684, 'GGA': 0.005263157894736842, 'GGG': 0.010526315789473684, 'GCA': 0.005263157894736842, 'UAC': 0.005263157894736842, 'CAC': 0.005263157894736842, 'UAG': 0.005263157894736842}\n"
     ]
    }
   ],
   "source": [
    "codon_counts = {}\n",
    "\n",
    "for i in range(0, len(reverse_complement_RNA_seq), 3):\n",
    "    \n",
    "    codon = reverse_complement_RNA_seq[i:(i + 3)]\n",
    "    \n",
    "    if codon in codon_counts:\n",
    "        codon_counts[codon] += 1\n",
    "    else:\n",
    "        codon_counts[codon] = 1\n",
    "        \n",
    "total_codons = sum(codon_counts.values())\n",
    "codon_freqs = {}\n",
    "\n",
    "for codon, codon_count in codon_counts.items():\n",
    "    codon_freqs[codon] = codon_count / total_codons\n",
    "    \n",
    "print('Codon frequencies: %s' % codon_freqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q3C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino-acid F (2 codons, 10 occurrences in the protein sequence):\n",
      "\tCodon UUU: 8 occurrences (80%)\n",
      "\tCodon UUC: 2 occurrences (20%)\n",
      "Amino-acid L (6 codons, 32 occurrences in the protein sequence):\n",
      "\tCodon CUA: 4 occurrences (12%)\n",
      "\tCodon CUC: 6 occurrences (18%)\n",
      "\tCodon UUG: 5 occurrences (15%)\n",
      "\tCodon CUU: 6 occurrences (18%)\n",
      "\tCodon CUG: 4 occurrences (12%)\n",
      "\tCodon UUA: 7 occurrences (21%)\n",
      "Amino-acid S (6 codons, 18 occurrences in the protein sequence):\n",
      "\tCodon AGU: 2 occurrences (11%)\n",
      "\tCodon AGC: 4 occurrences (22%)\n",
      "\tCodon UCG: 1 occurrences (5%)\n",
      "\tCodon UCU: 3 occurrences (16%)\n",
      "\tCodon UCA: 7 occurrences (38%)\n",
      "\tCodon UCC: 1 occurrences (5%)\n",
      "Amino-acid Y (2 codons, 3 occurrences in the protein sequence):\n",
      "\tCodon UAC: 1 occurrences (33%)\n",
      "\tCodon UAU: 2 occurrences (66%)\n",
      "Amino-acid C (2 codons, 14 occurrences in the protein sequence):\n",
      "\tCodon UGU: 6 occurrences (42%)\n",
      "\tCodon UGC: 8 occurrences (57%)\n",
      "Amino-acid P (4 codons, 19 occurrences in the protein sequence):\n",
      "\tCodon CCC: 4 occurrences (21%)\n",
      "\tCodon CCU: 9 occurrences (47%)\n",
      "\tCodon CCG: 0 occurrences (0%)\n",
      "\tCodon CCA: 6 occurrences (31%)\n",
      "Amino-acid H (2 codons, 3 occurrences in the protein sequence):\n",
      "\tCodon CAU: 2 occurrences (66%)\n",
      "\tCodon CAC: 1 occurrences (33%)\n",
      "Amino-acid Q (2 codons, 5 occurrences in the protein sequence):\n",
      "\tCodon CAG: 2 occurrences (40%)\n",
      "\tCodon CAA: 3 occurrences (60%)\n",
      "Amino-acid R (6 codons, 3 occurrences in the protein sequence):\n",
      "\tCodon CGU: 0 occurrences (0%)\n",
      "\tCodon CGG: 0 occurrences (0%)\n",
      "\tCodon CGA: 0 occurrences (0%)\n",
      "\tCodon CGC: 0 occurrences (0%)\n",
      "\tCodon AGG: 2 occurrences (66%)\n",
      "\tCodon AGA: 1 occurrences (33%)\n",
      "Amino-acid I (3 codons, 15 occurrences in the protein sequence):\n",
      "\tCodon AUU: 8 occurrences (53%)\n",
      "\tCodon AUC: 1 occurrences (6%)\n",
      "\tCodon AUA: 6 occurrences (40%)\n",
      "Amino-acid M (1 codons, 3 occurrences in the protein sequence):\n",
      "\tCodon AUG: 3 occurrences (100%)\n",
      "Amino-acid T (4 codons, 10 occurrences in the protein sequence):\n",
      "\tCodon ACG: 0 occurrences (0%)\n",
      "\tCodon ACC: 2 occurrences (20%)\n",
      "\tCodon ACA: 0 occurrences (0%)\n",
      "\tCodon ACU: 8 occurrences (80%)\n",
      "Amino-acid N (2 codons, 10 occurrences in the protein sequence):\n",
      "\tCodon AAU: 5 occurrences (50%)\n",
      "\tCodon AAC: 5 occurrences (50%)\n",
      "Amino-acid K (2 codons, 4 occurrences in the protein sequence):\n",
      "\tCodon AAG: 2 occurrences (50%)\n",
      "\tCodon AAA: 2 occurrences (50%)\n",
      "Amino-acid V (4 codons, 14 occurrences in the protein sequence):\n",
      "\tCodon GUU: 2 occurrences (14%)\n",
      "\tCodon GUC: 2 occurrences (14%)\n",
      "\tCodon GUG: 6 occurrences (42%)\n",
      "\tCodon GUA: 4 occurrences (28%)\n",
      "Amino-acid A (4 codons, 13 occurrences in the protein sequence):\n",
      "\tCodon GCA: 1 occurrences (7%)\n",
      "\tCodon GCU: 7 occurrences (53%)\n",
      "\tCodon GCC: 4 occurrences (30%)\n",
      "\tCodon GCG: 1 occurrences (7%)\n",
      "Amino-acid D (2 codons, 4 occurrences in the protein sequence):\n",
      "\tCodon GAU: 1 occurrences (25%)\n",
      "\tCodon GAC: 3 occurrences (75%)\n",
      "Amino-acid E (2 codons, 3 occurrences in the protein sequence):\n",
      "\tCodon GAA: 2 occurrences (66%)\n",
      "\tCodon GAG: 1 occurrences (33%)\n",
      "Amino-acid G (4 codons, 6 occurrences in the protein sequence):\n",
      "\tCodon GGC: 0 occurrences (0%)\n",
      "\tCodon GGU: 3 occurrences (50%)\n",
      "\tCodon GGA: 1 occurrences (16%)\n",
      "\tCodon GGG: 2 occurrences (33%)\n"
     ]
    }
   ],
   "source": [
    "# A dict mapping each aa into a set of its relevant codons\n",
    "aa_to_codons = {}\n",
    "\n",
    "for codon, aa in CODON_TABLE.items():\n",
    "    \n",
    "    if aa == '*':\n",
    "        continue\n",
    "    \n",
    "    if aa in aa_to_codons:\n",
    "        aa_to_codons[aa].add(codon)\n",
    "    else:\n",
    "        aa_to_codons[aa] = {codon}\n",
    "\n",
    "for aa, aa_codons in aa_to_codons.items():\n",
    "    \n",
    "    aa_count = protein_seq.count(aa)\n",
    "    \n",
    "    if aa_count == 0:\n",
    "        continue\n",
    "    \n",
    "    print('Amino-acid %s (%d codons, %d occurrences in the protein sequence):' % (aa, len(aa_codons), aa_count))\n",
    "    \n",
    "    for codon in aa_codons:\n",
    "        codon_count = codon_counts.get(codon, 0)\n",
    "        codon_freq = codon_count / aa_count\n",
    "        print('\\t' + 'Codon %s: %d occurrences (%d%%)' % (codon, codon_count, 100 * codon_freq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are clearly codons that appear more than others (within a given amino-acids). For example, of the 10 occurrences of F in the protein sequence, 8 (80%) are coded by the UUU codon, and only 2 (20%) by the UUC codon. To determine whether those deviations from a uniform distribution appear to be significant, we will need statistical tools (which we will only cover later in the course)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
