{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import csv\n",
    "import json\n",
    "from collections import defaultdict\n",
    "\n",
    "from Bio import SeqIO, pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.SubsMat.MatrixInfo import blosum62\n",
    "from Bio.SeqUtils import molecular_weight\n",
    "\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "# A patch required to let the csv library parsing long fields\n",
    "csv.field_size_limit(2 ** 31 - 1)\n",
    "\n",
    "DIR = r'c://downloads'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "685\n",
      "MNQSRSRSDGGSEETLPQDHNHHENERRWQQERLHREEAYYQFINELNDEDYRLMRDHNLLGTPGEITSEELQQRLDGVKEQLASQPDLRDGTNYRDSEVPRESSHEDSLLEWLNTFRRTGNATRSGQNGNQTWRAVSRTNPNNGEFRFSLEIHVNHENRGFEIHGEDYTDIPLSDSNRDHTANRQQRSTSPVARRTRSQTSVNFNGSSSNIPRTRLASRGQNPAEGSFSTLGRLRNGIGGAAGIPRANASRTNFSSHTNQSGGSELRQREGQRFGAAHVWENGARSNVTVRNTNQRLEPIRLRSTSNSRSRSPIQRQSGTVYHNSQRESRPVQQTTRRSVRRRGRTRVFLEQDRERERRGTAYTPFSNSRLVSRITVEEGEESSRSSTAVRRHPTITLDLQVRRIRPGENRDRDSIANRTRSRVGLAENTVTIESNSGGFRRTISRLERSGIRTYVSTITVPLRRISENELVEPSSVALRSILRQIMTGFGELSSLMEADSESELQRNGQHLPDMHSELSNLGTDNNRSQHREGSSQDRQAQGDSTEMHGENETTQPHTRNSDSRGGRQLRNPNNLVETGTLPILRLAHFFLLNESDDDDRIRGLTKEQIDNLSTRHYEHNSIDSELGKICSVCISDYVTGNKLRQLPCMHEFHIHCIDRWLSENCTCPICRQPVLGSNIANNG\n"
     ]
    }
   ],
   "source": [
    "# Parse the sequence of the human ring finger protein (C3H2C3 type) 6\n",
    "# Downloaded from UniProt at: https://www.uniprot.org/uniprot/A0A024RDP2.fasta\n",
    "human_record, = SeqIO.parse(os.path.join(DIR, 'A0A024RDP2.fasta'), 'fasta')\n",
    "human_seq = human_record.seq\n",
    "print(len(human_seq))\n",
    "print(human_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsed 384 ViralZone records.\n"
     ]
    }
   ],
   "source": [
    "# Parse the ViralZone records.\n",
    "\n",
    "viralzone_records = []\n",
    "\n",
    "with open(os.path.join(DIR, 'viralzone.csv'), 'r') as f:\n",
    "    \n",
    "    csv_reader = csv.reader(f)\n",
    "    \n",
    "    # Skip the header line.\n",
    "    next(csv_reader) \n",
    "    \n",
    "    for group, family, genus, strain, _, _, _, _, raw_genome in csv_reader:\n",
    "        viralzone_records.append({'group': group, 'family': family, 'genus': genus, 'strain': strain, \\\n",
    "                'genome': json.loads(raw_genome)})\n",
    "\n",
    "print('Parsed %d ViralZone records.' % len(viralzone_records))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 556 relevant proteins for the search.\n"
     ]
    }
   ],
   "source": [
    "# Extracting from ViralZone all the proteins from members of the Herpesviridae viral family that infect human (filtered by\n",
    "# demanding \"human\" to be in the virus's strain name).\n",
    "\n",
    "relevant_herpes_proteins = []\n",
    "\n",
    "for record in viralzone_records:\n",
    "    if record['family'] == 'Herpesviridae' and 'human' in record['strain'].lower():\n",
    "        for genome_segment in record['genome']:\n",
    "            for coding_region in genome_segment['coding_regions']:\n",
    "                relevant_herpes_proteins.append((coding_region['translation'], coding_region['product'], record['strain']))\n",
    "\n",
    "print('Extracted %d relevant proteins for the search.' % len(relevant_herpes_proteins))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best match: ubiquitin E3 ligase ICP0, of the viral strain Human herpesvirus 2 (score: 2.35)\n",
      "Lengths: human = 685, viral = 824\n",
      "Alignment:\n",
      "MNQSRSRSDGGSEETLPQDHNHHENERRWQQERLHREEAYYQFINELNDEDYRLMRDHNLLGTPGEITSEELQQRLDGVKEQLASQPDLRDGTNYRDSEVPRESSHEDSLLEWLNTFRRTGNATRSGQNGNQTWRAVSRTNPNNGEFRFSLEIHVNHENRGFEIHGEDYTDIPLSDSNRDHTANRQQRSTSPVARRTRSQTSVNFNGSSSNIPRTRLASRGQNPAEGSFSTLGRLRNGIGGAAGIPRANASRTNFSSHTNQSGGSELRQREGQRFGAAHVWENGARSNVTVRNTNQRLEPIRLRSTSNSRSRSPIQRQSGTVYHNSQRESRPVQQTTRRSVRRRGRTRVFLEQDRERERRGTAYTPFSNSRLVSRITVEEGEESSRSSTAVRRHPTITLDLQVRRIRPGENRDRDSIANRTRSRVGLAENTVTIESNSGGFRRTISRLERSGIRTYVSTITVPLRRISENELVEPSSVALRSILRQIMTGFGELSSLMEADSESELQRNGQHLPDMHSELSNLGTDNNRSQHREGSSQDRQAQGDSTEMHGENETTQPHTRNSDSRGGRQLRNPNNLVETGTLPILRLAHFFLLNESDDDDRIRGLTKEQIDNLSTRHYEHNSIDSELGKICSVCISDYVTGNKLRQLPCMHEFHIHCIDRWLSENCTCPICRQPVLGSNIANNG----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |..|.||.............||.|.|.|.|...|.....|||.|..||\n",
      "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------MEPRPGTSSRADPGPERPPRQTPGTPAAPHAWGMLNDMQWLASSDSEEETEVGISDDDLHRDSTSEAGSTDTEMFEAGLMDAATPPARPPAERQGSPTPADAQGSCGGGPVGEEEAEAGGGGDVCAVCTDEIAPPLRCQSFPCLHPFCIPCMKTWIPLRNTCPLCNTPVAYLIVGVTASGSFSTIPIVNDPRTRVEAEAAVRAGTAVDFIWTGNPRTAPRSLSLGGHTVRALSPTPPWPGTDDEDDDLADVDYVPPAPRRAPRRGGGGAGATRGTSQPAATRPAPPGAPRSSSSGGAPLRAGVGSGSGGGPAVAAVVPRVASLPPAAGGGRAQARRVGEDAAAAEGRTPPARQPRAAQEPPIVISDSPPPSPRRPAGPGPLSFVSSSSAQVSSGPGGGGLPQSSGRAARPRAAVAPRVRSPPRAAAAPVVSASADAAGPAPPAVPVDAHRAPRSRMTQAQTDTQAQSLGRAGATDARGSGGPGAEGGPGVPRGTNTPGAAPHAAEGAAARPRKRRGSDSGPAASSSASSSAAPRSPLAPQGVGAKRAAPRRAPDSDSGDRGHGPLAPASAGAAPPSASPSSQAAVAAASSSSASSSSASSSSASSSSASSSSASSSSASSSSASSSAGGAGGSVASASGAGERRETSLGPRAAAPRGPRKCARKTRHAEGGPEPGARDPAPGLTRYLPIAGVSSVVALAPYVNKTVTGDCLPVLDMETGHIGAYVVLVDQTGNVADLLRAAAPAWSRRTLLPEHARNCVRPPDYPTPPASEWNSLWMTPVGNMLFDQGTLVGALDFHGLRSRHPWSREQGAPAPAGDAPAGHGE\n",
      "  Score=113\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Among the candidates, find the viral protein with the best match.\n",
    "# Note that we choose to run a local alignment, because we want to allow distant homologues (including small pieces hijacked\n",
    "# by the virus).\n",
    "# Like in the lab exercise, we normalize the score by the length of the alignment, but we require it to be not too short (as\n",
    "# an arbitrary choice, at least 30 residues).\n",
    "\n",
    "best_score = None\n",
    "best_match = None\n",
    "\n",
    "for i, (seq, protein_name, viral_strain) in enumerate(relevant_herpes_proteins):\n",
    "    \n",
    "    print('%d/%d' % (i + 1, len(relevant_herpes_proteins)), end = '\\r')\n",
    "\n",
    "    (_, _, score, begin, end), = pairwise2.align.localds(human_seq, seq, blosum62, -10.0, -0.5, one_alignment_only = True)\n",
    "    alignment_length = end - begin\n",
    "    \n",
    "    if alignment_length < 30:\n",
    "        continue\n",
    "    \n",
    "    score /= alignment_length\n",
    "    \n",
    "    if best_score is None or best_score < score:\n",
    "        best_score = score\n",
    "        best_match = (seq, protein_name, viral_strain)\n",
    "\n",
    "seq, protein_name, viral_strain = best_match\n",
    "print('Best match: %s, of the viral strain %s (score: %.2f)' % (protein_name, viral_strain, best_score))\n",
    "print('Lengths: human = %d, viral = %d' % (len(human_seq), len(seq)))\n",
    "print('Alignment:')\n",
    "alignment, = pairwise2.align.localds(human_seq, seq, blosum62, -10.0, -0.5, one_alignment_only = True)\n",
    "print(format_alignment(*alignment))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Virus-human homologues and, more generally, virus-host homologues, sometimes occur through the process of horizontal gene transfer. Such horizontal transfer can occur in both directions (i.e. either pieces of the viral DNA are integrated into the host's genome, or the virus hijacks genes from the host). See for example: [Liu, H. et al. Widespread horizontal gene transfer from double-stranded RNA viruses to eukaryotic nuclear genomes. J. Virol. 84, 11876â€“11887 (2010).](https://jvi.asm.org/content/84/22/11876.short)\n",
    "\n",
    "In this specific case, the homology we have found seems quite weak. If I was forced to make a guess, I'd say that the homology does seem genuine, but I am far from certain (a more educated judgement would require more advanced statistical tools and evolutionary models, and/or better expertise with the subject matter).\n",
    "\n",
    "The reasons that I tend to accept this homology are:\n",
    "1. The local alignment is continuous (without any gaps in it).\n",
    "2. There are many exact matches in the shared segment (17 of 48 residues, within proteins of 685 and 824 amino-acids).\n",
    "3. Even among the mismatches, many are reasonable amino-acid pairs (e.g. K/R, D/E).\n",
    "\n",
    "On the other hand, we need to keep in mind that we have attempted 556 protein pairs, searching for the best local alignment within each pair, so some measure of similarity is expected even at random."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q2\n",
    "\n",
    "About whether the researchers' conclusion is justified - hell no! The human genome is far larger than the E. coli genome (~3 billion compared to ~5 million bp - a three orders-of-magnitude difference!), so it is expected that any random sequence would find a better match to some segment of the human genome than of the E. coli genome. Another major problem stems from the fact that we are talking about a meta-genomic sample, which is likely a mix of many organisms and viruse. In reality, the sample is likely comprised of both eukaryotes AND prokaryotes (so a meaningful conclusion from this analysis could be, for example, an estimate of the relative abundance of organisms belonging to each of the two domains in the sequenced sample).\n",
    "\n",
    "If we want this analysis to make sense, it should be accompanied by some sort of probabilistic calculation that accounts for the different genome lengths of the two species. It would also be better to compare the seqeunces to all prokaryote and eukaryote organisms (instead of just human and E. coli as two arbitary representatives)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Correlations between relative mass to any of the three 2D structures:\n",
      "\tHelix: r = 0.23, p-value = 3.59e-72\n",
      "\tStrand: r = 0.03, p-value = 3.24e-02\n",
      "\tTurn: r = 0.13, p-value = 8.05e-24\n"
     ]
    }
   ],
   "source": [
    "# We iterate over all reviewed UniProt human proteins, extracting the total number of residues for each of the three secondary structures, and\n",
    "# calculating the molecular weight of each protein from the reviewed human proteins downloaded from UniProt.\n",
    "\n",
    "SS_OPTIONS = [\n",
    "    'helix',\n",
    "    'strand',\n",
    "    'turn',\n",
    "]\n",
    "\n",
    "def get_total_feature_length(uniprot_record, feature_type):\n",
    "    return sum([feature.location.end - feature.location.start for feature in uniprot_record.features if \\\n",
    "            feature.type == feature_type])\n",
    "\n",
    "relative_masses = []\n",
    "relative_lengths_per_ss = defaultdict(list)\n",
    "\n",
    "# From: http://www.uniprot.org/uniprot/?query=organism%3A%22Homo+sapiens+%5B9606%5D%22+AND+reviewed%3Ayes&sort=score\n",
    "with gzip.open(os.path.join(DIR, 'uniprot_human_reviewed.xml.gz'), 'r') as f:\n",
    "    for i, record in enumerate(SeqIO.parse(f, 'uniprot-xml')):\n",
    "    \n",
    "        print(i, end = '\\r')\n",
    "        record_relative_lengths_per_ss = {ss: get_total_feature_length(record, ss) / len(record.seq) for ss in SS_OPTIONS}\n",
    "        \n",
    "        # If there are no residues with any secondary structure (i.e. the sum of the relative lengths, over all secondary\n",
    "        # structures, is 0), then it the protein is excluded from the analysis (it probably just has missing annotations).\n",
    "        if sum(record_relative_lengths_per_ss.values()) > 0:\n",
    "            \n",
    "            for ss, record_ss_relative_length in record_relative_lengths_per_ss.items():\n",
    "                relative_lengths_per_ss[ss].append(record_ss_relative_length)\n",
    "\n",
    "            relative_masses.append(molecular_weight(record.seq, seq_type = 'protein') / len(record.seq))\n",
    "            \n",
    "print('Correlations between relative mass to any of the three 2D structures:')\n",
    "            \n",
    "for ss, ss_relative_lengths in relative_lengths_per_ss.items():\n",
    "    print('\\t' + '%s: r = %.2f, p-value = %.2e' % ((ss.capitalize(),) + pearsonr(relative_masses, ss_relative_lengths)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All three secondary structures give a significant, positive correlation with molecular weight (though in the case of Strand, the significance is borderline). The correlation is strongest for Helix (Pearson's coefficient of 0.23), and most significant (a p-value of 3.6e-72 cannot be attributed to chance). We conclude that molecular weight is in fact associated with secondary structure, but whether this is considered \"substantial\" is a matter of definition (personally, I would consider it a medium strength association for Helix and Turn, and weak, if any, for Strand). "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
