{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "from collections import defaultdict\n",
    "\n",
    "DIR = r'c://downloads'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "CODON_TABLE = {\n",
    "    'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "    'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
    "    'UAU': 'Y', 'UAC': 'Y', 'UAA': '*', 'UAG': '*',\n",
    "    'UGU': 'C', 'UGC': 'C', 'UGA': '*', 'UGG': 'W',\n",
    "    'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "    'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
    "    'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
    "    'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'GAU': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G',\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let us parse the file, like in the lab exercise.\n",
    "\n",
    "f = open(os.path.join(DIR, 'orf_exons_chr17.txt'), 'r')\n",
    "\n",
    "last_gene = None\n",
    "orf_exons_per_gene = {}\n",
    "\n",
    "for line in f:\n",
    "    \n",
    "    line = line.strip()\n",
    "    \n",
    "    if line.startswith('ORF Exon #'):\n",
    "        orf_exon_seq = line[(line.find(': ') + 2):]\n",
    "        orf_exons_per_gene[last_gene].append(orf_exon_seq)\n",
    "    else:\n",
    "        last_gene = line.replace(':', '')\n",
    "        orf_exons_per_gene[last_gene] = []\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We will use the data we parsed to reconstruct the coding DNA sequence of each gene (i.e. the part of the gene's sequence \n",
    "# coding to protein, which corresponds to the mRNA transcript without the UTRs), and use that CDS sequence to reconstruct\n",
    "# the protein sequence.\n",
    "\n",
    "def translate_seq(rna_seq):\n",
    "    \n",
    "    aa_letters = []\n",
    "    \n",
    "    for i in range(0, len(rna_seq), 3):\n",
    "        codon = rna_seq[i:(i + 3)]\n",
    "        aa_letters.append(CODON_TABLE[codon])\n",
    "        \n",
    "    return ''.join(aa_letters)\n",
    "\n",
    "dna_seq_per_gene = {}\n",
    "protein_seq_per_gene = {}\n",
    "\n",
    "for gene_symbol, orf_exons in orf_exons_per_gene.items():\n",
    "    dna_seq = ''.join(orf_exons)\n",
    "    dna_seq_per_gene[gene_symbol] = dna_seq\n",
    "    protein_seq = translate_seq(dna_seq.replace('T', 'U'))\n",
    "    protein_seq_per_gene[gene_symbol] = protein_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FZD2\n",
      "GNA13\n",
      "GRIN2C\n",
      "ANKRD13B\n",
      "SOST\n",
      "UTS2R\n"
     ]
    }
   ],
   "source": [
    "# Find genes on chr17 with GCGCGCGCGC in their coding DNA sequence.\n",
    "\n",
    "for gene_symbol, dna_seq in dna_seq_per_gene.items():\n",
    "    if 'GCGCGCGCGC' in dna_seq:\n",
    "        print(gene_symbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EFCAB3\n"
     ]
    }
   ],
   "source": [
    "# Find protein-coding genes on chr17 with RKRKRK in their protein sequence.\n",
    "\n",
    "for gene_symbol, protein_seq in protein_seq_per_gene.items():\n",
    "    if 'RKRKRK' in protein_seq:\n",
    "        print(gene_symbol)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q2A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stop codons: UGA, UAG, UAA\n"
     ]
    }
   ],
   "source": [
    "# Extract the three stop codons from the codon table (of course a quick Google search would also work).\n",
    "\n",
    "stop_codons = set()\n",
    "\n",
    "for codon, aa in CODON_TABLE.items():\n",
    "    if aa == '*':\n",
    "        stop_codons.add(codon)\n",
    "        \n",
    "print('Stop codons: %s' % ', '.join(stop_codons))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CACUAUAGCCGGUCGUACCCUAAGGCAGAAGCAUAUCGGCGACGCUGGAUAGCCCAUGCGAAAGUACCCCUUCUAACUGUAAGGCUACUCGAAGAGUAA'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_LETTERS = list('ACGU')\n",
    "\n",
    "def get_random_rna_nt():\n",
    "    return random.choice(RNA_LETTERS)\n",
    "\n",
    "def generate_random_codon():\n",
    "    return get_random_rna_nt() + get_random_rna_nt() + get_random_rna_nt()\n",
    "\n",
    "def generate_random_transcript():\n",
    "    \n",
    "    generated_codons = []\n",
    "    \n",
    "    while True:\n",
    "        \n",
    "        codon = generate_random_codon()\n",
    "        generated_codons.append(codon)\n",
    "        \n",
    "        if codon in stop_codons:\n",
    "            break\n",
    "            \n",
    "    return ''.join(generated_codons)\n",
    "            \n",
    "generate_random_transcript()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q2B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average trasncript length of 1,000 randomly generated RNA transcripts: 64.59 nt.\n"
     ]
    }
   ],
   "source": [
    "generated_transcript_lengths = []\n",
    "\n",
    "for _ in range(1000):\n",
    "    generated_transcript_lengths.append(len(generate_random_transcript()))\n",
    "    \n",
    "print('Average trasncript length of 1,000 randomly generated RNA transcripts: %.2f nt.' % (sum(generated_transcript_lengths) / \\\n",
    "        len(generated_transcript_lengths)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q2C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcripts with a GC content of approximately 10% [225 transcripts]: average length is 10.47 nt.\n",
      "Transcripts with a GC content of approximately 20% [2312 transcripts]: average length is 10.16 nt.\n",
      "Transcripts with a GC content of approximately 30% [9423 transcripts]: average length is 11.94 nt.\n",
      "Transcripts with a GC content of approximately 40% [18347 transcripts]: average length is 47.19 nt.\n",
      "Transcripts with a GC content of approximately 50% [48973 transcripts]: average length is 88.14 nt.\n",
      "Transcripts with a GC content of approximately 60% [16525 transcripts]: average length is 63.57 nt.\n",
      "Transcripts with a GC content of approximately 70% [2250 transcripts]: average length is 20.49 nt.\n",
      "Transcripts with a GC content of approximately 80% [173 transcripts]: average length is 14.45 nt.\n"
     ]
    }
   ],
   "source": [
    "transcript_lengths_per_gc_content = defaultdict(list)\n",
    "\n",
    "for _ in range(100000):\n",
    "    \n",
    "    random_transcript_seq = generate_random_transcript()\n",
    "    transcript_length = len(random_transcript_seq)\n",
    "    gc_content = (random_transcript_seq.count('G') + random_transcript_seq.count('C')) / transcript_length\n",
    "    # Rounding the GC content to a resolution of 1 digit will provide its proper bin (e.g. ~30%).\n",
    "    gc_content_group = round(gc_content, 1)\n",
    "    \n",
    "    if 0.1 <= gc_content_group <= 0.9:\n",
    "        transcript_lengths_per_gc_content[gc_content_group].append(transcript_length)\n",
    "    \n",
    "for gc_content, transcript_lengths in sorted(transcript_lengths_per_gc_content.items()):\n",
    "    n_transcripts = len(transcript_lengths)\n",
    "    avg_length = sum(transcript_lengths) / n_transcripts\n",
    "    print('Transcripts with a GC content of approximately %d%% [%d transcripts]: average length is %.2f nt.' % \\\n",
    "            (100 * gc_content, n_transcripts, avg_length))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that average transcript length indeed varies, quite substantially, between GC-content groups.\n",
    "\n",
    "The more obious reason for a (random) transcript length to depend on GC content is that stop codons are enriched with A/U compared to G/C (in the three stop codons, A/U occur 7 times, while G/C only 2 times - that is quite an imbalance). In short transcripts, the stop codon comprises a larger portion of all the the transcript's nucleotides, so the transcript will end up more enriched with A/U (as the transcript's stop codon is enriched with A/U). In longer transcripts, the imbalance of the stop codon would be less substantial in the overall transcript's GC content, so this effect will not be so noticeable. To summarize this effect: the imblance of GC content in stop codons is expected to lead to transcripts with a higher GC content to be, on average, longer.\n",
    "\n",
    "What we get in reality, however, is more complicated than that. Transcripts with low GC content are, as expected, very short (e.g. look at the transcripts with a GC content of ~10%). But transcripts with a very high GC content (e.g. ~90%) are also short. What's going on? Well, there's another force at play here. In the process of collecting transcripts with a given GC content, we generated transcripts in an unbiased way (letting each of the 4 nucleotides occur with 25% probability), and collected them into GC-content groups only post-hoc. In other words, the collection of transcripts of the 90% group ended up whatever transcripts that ended up with a GC content of ~90% (in probability theory this is called conditioning). In the case of GC contents that are very extreme (e.g. 10% of 90%), this creates a bias towards very short transcripts, because long transcripts are very unlikely to end up with an imbalance of 10% or 90% (for short transcripts this is much more plausible, due to the \"Law of small numbers\").\n",
    "\n",
    "To summarize, we have two forces at play here. First, high GC content naturally leads to longer transcripts (due to the low GC content of stop codons). But more importantly, very extreme GC contents (e.g. 10% or 90%) bias the resulst towards shorter transcripts. We get that transcripts with very high GC content also end up quite short (but not as short as those with a very low GC content; e.g. note that although the 90% transcripts are short, the 10% are even shorter). Since the first phenomenon is more substantial, we end up with the longest transcripts being those with a balanced GC content (~50%)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
