{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import csv\n",
    "from collections import defaultdict, Counter\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "DIR = r'c://downloads'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Translations do not match for: NM_001350809.1\n",
      "Translations do not match for: NM_001317231.1\n",
      "Translations do not match for: NM_001362.3\n",
      "Translations do not match for: NM_182704.1\n",
      "Translations do not match for: NM_021237.5\n",
      "Translations do not match for: NM_001319071.2\n",
      "Translations do not match for: NM_001316374.2\n",
      "Translations do not match for: NM_206926.2\n",
      "Translations do not match for: NM_001318069.2\n",
      "Translations do not match for: NM_001166304.2\n",
      "Translations do not match for: NM_001173513.2\n",
      "Translations do not match for: NM_001017371.5\n",
      "Translations do not match for: NM_182701.1\n",
      "Translations do not match for: NM_016332.3\n",
      "Translations do not match for: NM_001352301.1\n",
      "Translations do not match for: NM_001352302.1\n",
      "Translations do not match for: NM_006440.5\n",
      "Translations do not match for: NM_001099456.3\n",
      "Translations do not match for: NM_014293.4\n",
      "Translations do not match for: NM_052883.2\n",
      "Translations do not match for: NM_001352300.2\n",
      "Translations do not match for: NM_199072.4\n",
      "Translations do not match for: NM_024424.5\n",
      "Translations do not match for: NM_001290266.1\n",
      "Translations do not match for: NM_001085386.2\n",
      "Translations do not match for: NM_001166346.1\n",
      "Translations do not match for: NM_002097.3\n",
      "Processed 32200 transcripts.\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Nadav\\Anaconda2\\envs\\py37\\lib\\site-packages\\Bio\\Seq.py:2576: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Translations do not match for: NM_001301020.1\n",
      "Translations do not match for: NM_004152.3\n",
      "Translations do not match for: NM_001301302.1\n",
      "Translations do not match for: NM_001100163.2\n",
      "Translations do not match for: NM_002537.3\n",
      "Translations do not match for: NM_001357016.1\n",
      "Translations do not match for: NM_003214.4\n",
      "Translations do not match for: NM_001042559.2\n",
      "Translations do not match for: NM_002083.4\n",
      "Translations do not match for: NM_003009.4\n",
      "Translations do not match for: NM_001172705.1\n",
      "Translations do not match for: NM_001418.4\n",
      "Translations do not match for: NM_080430.4\n",
      "Translations do not match for: NM_175886.3\n",
      "Translations do not match for: NM_182742.3\n",
      "Translations do not match for: NM_182729.3\n",
      "Translations do not match for: NM_003330.4\n",
      "Translations do not match for: NM_182743.3\n",
      "Translations do not match for: NM_001261446.2\n",
      "Translations do not match for: NM_001261445.2\n",
      "Translations do not match for: NM_017861.4\n",
      "Translations do not match for: NM_031895.6\n",
      "Translations do not match for: NM_004261.5\n",
      "Translations do not match for: NM_203341.2\n",
      "Translations do not match for: NM_001128619.3\n",
      "Translations do not match for: NM_001364289.1\n",
      "Translations do not match for: NM_001364287.1\n",
      "Translations do not match for: NM_001317122.1\n",
      "Translations do not match for: NM_001317384.2\n",
      "Translations do not match for: NM_001093771.3\n",
      "Translations do not match for: NM_020451.3\n",
      "Translations do not match for: NM_033505.4\n",
      "Translations do not match for: NM_153756.3\n",
      "Translations do not match for: NM_006386.5\n",
      "Translations do not match for: NM_001098504.2\n",
      "Translations do not match for: NM_001366496.1\n",
      "Translations do not match for: NM_000793.6\n",
      "Translations do not match for: NM_001324462.2\n",
      "Translations do not match for: NM_001329790.2\n",
      "Translations do not match for: NM_000792.7\n",
      "Translations do not match for: NM_001039716.3\n",
      "Translations do not match for: NM_201441.3\n",
      "Translations do not match for: NM_002084.5\n",
      "Translations do not match for: NM_003213.4\n",
      "Translations do not match for: NM_213593.5\n",
      "Translations do not match for: NM_001172437.2\n",
      "Translations do not match for: NM_001184962.2\n",
      "Translations do not match for: NM_033013.2\n",
      "Translations do not match for: NM_003889.3\n",
      "Translations do not match for: NM_000581.4\n",
      "Translations do not match for: NM_001184961.1\n",
      "Translations do not match for: NM_015068.3\n",
      "Translations do not match for: NM_016275.5\n",
      "Translations do not match for: NM_001315491.2\n",
      "Translations do not match for: NM_001329503.2\n",
      "Translations do not match for: NM_001329502.2\n",
      "Translations do not match for: NM_201397.3\n",
      "Translations do not match for: NM_004323.6\n",
      "Translations do not match for: NM_021182.3\n",
      "Translations do not match for: NM_001171940.2\n",
      "Translations do not match for: NM_170746.4\n",
      "Translations do not match for: NM_018445.6\n",
      "Translations do not match for: NM_001321335.1\n",
      "Translations do not match for: NM_001315537.1\n",
      "Translations do not match for: NM_024426.6\n",
      "Translations do not match for: NM_001134939.1\n",
      "Translations do not match for: NM_016178.2\n",
      "Translations do not match for: NM_001301371.1\n",
      "Translations do not match for: NM_000378.6\n",
      "Translations do not match for: NM_002110.4\n",
      "Translations do not match for: NM_001172130.2\n",
      "Translations do not match for: NM_021961.6\n",
      "Translations do not match for: NM_013989.5\n",
      "Translations do not match for: NM_031454.2\n",
      "Translations do not match for: NM_012248.4\n",
      "Translations do not match for: NM_002085.5\n",
      "Translations do not match for: NM_001039847.3\n",
      "Translations do not match for: NM_001367832.1\n",
      "Translations do not match for: NM_001039848.4\n",
      "Translations do not match for: NM_001318497.2\n",
      "Translations do not match for: NM_020860.4\n",
      "Translations do not match for: NM_001169117.2\n",
      "Translations do not match for: NM_001169118.2\n",
      "Translations do not match for: NM_002006.5\n",
      "Translations do not match for: NM_203289.5\n",
      "Translations do not match for: NM_001173531.2\n",
      "Translations do not match for: NM_001318853.2\n",
      "Translations do not match for: NM_018646.6\n",
      "Translations do not match for: NM_001287424.2\n",
      "Translations do not match for: NM_001364085.2\n",
      "Translations do not match for: NM_003376.6\n",
      "Translations do not match for: NM_005410.4\n",
      "Translations do not match for: NM_001093726.2\n",
      "Translations do not match for: NM_001025369.3\n",
      "Translations do not match for: NM_001025368.3\n",
      "Translations do not match for: NM_001025366.3\n",
      "Translations do not match for: NM_001085486.2\n",
      "Translations do not match for: NM_001025367.3\n",
      "Translations do not match for: NM_001204385.2\n",
      "Translations do not match for: NM_001317010.1\n",
      "Translations do not match for: NM_001025370.3\n",
      "Translations do not match for: NM_001171622.2\n",
      "Translations do not match for: NM_001033756.3\n",
      "Translations do not match for: NM_001354870.1\n",
      "Translations do not match for: NM_001243186.2\n",
      "Translations do not match for: NM_020831.6\n",
      "Translations do not match for: NM_002467.6\n",
      "Translations do not match for: NM_001282662.3\n",
      "Translations do not match for: NM_001282661.3\n",
      "**************************************************\n",
      "5_utr\n",
      "\tAvg. relative length: 8%\n",
      "\tAvg. GC content: 61%\n",
      "cds\n",
      "\tAvg. relative length: 49%\n",
      "\tAvg. GC content: 51%\n",
      "3_utr\n",
      "\tAvg. relative length: 41%\n",
      "\tAvg. GC content: 42%\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "We download the entire repertoire of human transcriptome from NCBI in the genebank format (the human.#.rna.gbff.gz files at\n",
    "ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/mRNA_Prot/). We iterate over these transcripts and use the \"CDS\" feature in order\n",
    "to determine that the trasncript is truly protein coding, and to find its 5' UTR, CDS and 3' UTR regions. We further validate\n",
    "that the obtained CDS sequence truly translates to the amino-acid sequence that appears in the record as \"translation\".\n",
    "Additionally, we make sure that the Comment of the record starts with the word \"REVIEWED\", to make sure we only consider\n",
    "reviewed records.\n",
    "Running this code on the entire human transcriptome is expected to take a few hours...\n",
    "'''\n",
    "\n",
    "def get_features_of_type(features, type_name):\n",
    "    return [feature for feature in features if feature.type == type_name]\n",
    "    \n",
    "# Mapping region names ('5_utr', 'cds' and '3_utr') to the observed relative lengths and GC contents of the transcripts.\n",
    "relative_lengths = defaultdict(list)\n",
    "gc_contents = defaultdict(list)\n",
    "\n",
    "transcript_count = 0\n",
    "\n",
    "for file_index in range(8):\n",
    "\n",
    "    # From: ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/mRNA_Prot/\n",
    "    f = gzip.open(os.path.join(DIR, 'human.%d.rna.gbff.gz' % (file_index + 1)), 'rt')\n",
    "\n",
    "    for record in SeqIO.parse(f, 'genbank'):\n",
    "    \n",
    "        transcript_count += 1\n",
    "        \n",
    "        if transcript_count % 100 == 0:\n",
    "            print('Processed %d transcripts.' % transcript_count, end = '\\r')\n",
    "\n",
    "        cds_features = get_features_of_type(record.features, 'CDS')\n",
    "\n",
    "        if len(cds_features) > 0 and record.annotations['comment'].lower().startswith('reviewed'):\n",
    "\n",
    "            cds_feature, = cds_features\n",
    "            cds_start, cds_end = int(cds_feature.location.start), int(cds_feature.location.end)\n",
    "            translation, = cds_feature.qualifiers['translation']\n",
    "\n",
    "            seqs = {\n",
    "                '5_utr': record.seq[:cds_start],\n",
    "                'cds': record.seq[cds_start:cds_end],\n",
    "                '3_utr': record.seq[cds_end:],\n",
    "            }\n",
    "\n",
    "            cds_translation = seqs['cds'].translate()\n",
    "\n",
    "            if not cds_translation.endswith('*') or translation != cds_translation[:-1]:\n",
    "                print('Translations do not match for: ' + record.id)\n",
    "                continue\n",
    "\n",
    "            for region, seq in seqs.items():\n",
    "                relative_lengths[region].append(len(seq) / len(record.seq))\n",
    "                gc_contents[region].append(GC(seq))\n",
    "                \n",
    "    f.close()\n",
    "            \n",
    "print('*' * 50)\n",
    "\n",
    "for region in relative_lengths:\n",
    "    print(region)\n",
    "    print('\\t' + 'Avg. relative length: %d%%' % (100 * sum(relative_lengths[region]) / len(relative_lengths[region])))\n",
    "    print('\\t' + 'Avg. GC content: %d%%' % (sum(gc_contents[region]) / len(gc_contents[region])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intron beginning pairs:\n",
      "\tGT: 98.16%\n",
      "\tGC: 1.39%\n",
      "\tAT: 0.11%\n",
      "\tTG: 0.05%\n",
      "\tAG: 0.04%\n",
      "\tGG: 0.03%\n",
      "\tCT: 0.03%\n",
      "\tGA: 0.03%\n",
      "\tAA: 0.03%\n",
      "\tTA: 0.03%\n",
      "\tTT: 0.02%\n",
      "\tCA: 0.02%\n",
      "\tCC: 0.02%\n",
      "\tAC: 0.02%\n",
      "\tT: 0.01%\n",
      "\tTC: 0.01%\n",
      "\tC: 0.01%\n",
      "\tA: 0.01%\n",
      "\tCG: 0.00%\n",
      "\tG: 0.00%\n",
      "Intron end pairs:\n",
      "\tAG: 99.46%\n",
      "\tAC: 0.10%\n",
      "\tAA: 0.06%\n",
      "\tGG: 0.05%\n",
      "\tCC: 0.05%\n",
      "\tTG: 0.05%\n",
      "\tCT: 0.03%\n",
      "\tAT: 0.03%\n",
      "\tTT: 0.03%\n",
      "\tGA: 0.02%\n",
      "\tTC: 0.02%\n",
      "\tGC: 0.02%\n",
      "\tTA: 0.02%\n",
      "\tGT: 0.02%\n",
      "\tCA: 0.02%\n",
      "\tCG: 0.01%\n",
      "\tT: 0.01%\n",
      "\tC: 0.01%\n",
      "\tA: 0.01%\n",
      "\tG: 0.00%\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "We download and extract the entire seqeunce of chromosome 11 from the FASTA file found at:\n",
    "ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/chr11.fa.gz.\n",
    "Then, we use GENCODE annotations to find all exon annotations on chromosome 11, and group them by transcript. The grouping by\n",
    "transcript is very important, because we want only intronic regions (not intergenic).\n",
    "For each transcript we then use the coordinates of its exons to find the coordinates of the introns, and we extract the relevant\n",
    "sequence from the chr11 sequence to get the intron seqeunces. We then use these sequences to extract only the pair of letter\n",
    "at the beginning and end of each intron.\n",
    "Note that it is crucial to use the same version of the human reference genome (GRCh38) for both files.\n",
    "'''\n",
    "\n",
    "# ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/chr11.fa.gz\n",
    "f = gzip.open(os.path.join(DIR, 'chr11.fa.gz'), 'rt')\n",
    "chr11_record, = SeqIO.parse(f, 'fasta')\n",
    "chr11_seq = chr11_record.seq\n",
    "f.close()\n",
    "\n",
    "# ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_29/gencode.v29.annotation.gtf.gz\n",
    "f = gzip.open(os.path.join(DIR, 'gencode.v29.annotation.gtf.gz'), 'rt')\n",
    "csv_reader = csv.reader(f, delimiter = '\\t')\n",
    "\n",
    "for _ in range(5):\n",
    "    next(csv_reader)\n",
    "\n",
    "def parse_extra_fields(raw_extra_fields):\n",
    "\n",
    "    extra_fields = {}\n",
    "\n",
    "    for raw_extra_field in raw_extra_fields[:-1].split(';'):\n",
    "        key, raw_value = raw_extra_field.strip().split(' ')\n",
    "        value = raw_value.strip('\"')\n",
    "        extra_fields[key] = value\n",
    "        \n",
    "    return extra_fields\n",
    "    \n",
    "exons_per_transcript = defaultdict(list)\n",
    "    \n",
    "for a_chr, _, a_type, a_start, a_end, _, strand, _, raw_extra_fields in csv_reader:\n",
    "    if a_chr == 'chr11' and a_type == 'exon':\n",
    "        extra_fields = parse_extra_fields(raw_extra_fields)\n",
    "        exons_per_transcript[extra_fields['transcript_id']].append((strand, int(a_start), int(a_end)))\n",
    "    \n",
    "f.close()\n",
    "\n",
    "start_pairs = Counter()\n",
    "end_pairs = Counter()\n",
    "\n",
    "for transcript_exons in exons_per_transcript.values():\n",
    "    \n",
    "    strand, = {strand for strand, start, end in transcript_exons}\n",
    "    \n",
    "    if strand == '+':\n",
    "        # On the positive strand, introns are between the end of the last exon (transcript_exons[i][2]) to the start of the next\n",
    "        # one (transcript_exons[i + 1][1])\n",
    "        introns = [(transcript_exons[i][2] + 1, transcript_exons[i + 1][1] - 1) for i in range(len(transcript_exons) - 1)]\n",
    "    else:\n",
    "        # On the negative strand, the exons appear in reverse order with respect to the chromosome coordinates (they are in\n",
    "        # the correct order with respect to the gene itself). Therefore, each intron is found between the end of the next\n",
    "        # exon (transcript_exons[i + 1][2]) to the start of the last one (transcript_exons[i][1]).\n",
    "        introns = [(transcript_exons[i + 1][2] + 1, transcript_exons[i][1] - 1) for i in range(len(transcript_exons) - 1)]\n",
    "        \n",
    "    for start, end in introns:\n",
    "        \n",
    "        intron_seq = chr11_seq[(start - 1):end]\n",
    "        \n",
    "        if strand == '-':\n",
    "            intron_seq = intron_seq.reverse_complement()\n",
    "            \n",
    "        start_pairs[str(intron_seq[:2]).upper()] += 1\n",
    "        end_pairs[str(intron_seq[-2:]).upper()] += 1\n",
    "        \n",
    "print('Intron beginning pairs:')\n",
    "\n",
    "for pair, count in start_pairs.most_common():\n",
    "    print('\\t' + '%s: %.2f%%' % (pair, 100 * count / sum(start_pairs.values())))\n",
    "\n",
    "print('Intron end pairs:')\n",
    "\n",
    "for pair, count in end_pairs.most_common():\n",
    "    print('\\t' + '%s: %.2f%%' % (pair, 100 * count / sum(end_pairs.values())))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The major spliceosome (constituting ~99% of splicing) splices introns containing GU at the 5' splice site and AG at the 3' splice site (these are known as the \"canonical splice sites\"; see [RNA splicing](https://en.wikipedia.org/wiki/RNA_splicing)). Our results, then, appear to be in line with lieterature."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
