{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "DIR = r'c://downloads'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# From: https://www.ncbi.nlm.nih.gov/nuccore/NM_000518 (download as FASTA)\n",
    "hbb_record, = SeqIO.parse(os.path.join(DIR, 'sequence.fasta'), 'fasta')\n",
    "hbb_seq = hbb_record.seq\n",
    "hbb_translated_seq = hbb_seq[50:494].translate()[:-1]\n",
    "\n",
    "hbb_protein_record = SeqRecord(hbb_translated_seq, id = hbb_record.id, description = hbb_record.description)\n",
    "SeqIO.write(hbb_protein_record, os.path.join(DIR, 'translation.fasta'), 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The longest record is NM_001267550.2 [length: 109224 nt]\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "\n",
    "longest_record = None\n",
    "\n",
    "for i in range(8):\n",
    "    \n",
    "    # Downloaded the 8 files From: ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/mRNA_Prot/\n",
    "    f = gzip.open(os.path.join(DIR, 'human.%d.rna.fna.gz' % (i + 1)), 'rt')\n",
    "    records = SeqIO.parse(f, 'fasta')\n",
    "    \n",
    "    for record in records:\n",
    "        if longest_record is None or len(record.seq) > len(longest_record.seq):\n",
    "            longest_record = record\n",
    "\n",
    "    f.close()\n",
    "    \n",
    "print('The longest record is %s [length: %d nt]' % (longest_record.id, len(longest_record.seq)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44.07364681754925\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqUtils import GC\n",
    "print(GC(longest_record.seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 363 exons.\n",
      "Max GC content: 59.52\n"
     ]
    }
   ],
   "source": [
    "# From: https://www.ncbi.nlm.nih.gov/nuccore/NM_001267550\n",
    "record, = SeqIO.parse(os.path.join(DIR, 'NM_001267550.gb'), 'genbank')\n",
    "exons = [(int(feature.location.start), int(feature.location.end)) for feature in record.features if feature.type == 'exon']\n",
    "print('There are %d exons.' % len(exons))\n",
    "\n",
    "exon_seqs = [record.seq[start:end] for start, end in exons]\n",
    "print('Max GC content: %.2f' % max(map(GC, exon_seqs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M': 0.011336167375177128, 'T': 0.07257369898030062, 'Q': 0.027367953099385958, 'A': 0.06143202467283487, 'P': 0.0797699424856214, 'F': 0.025756439109777443, 'L': 0.06029285099052541, 'S': 0.07015642799588787, 'V': 0.09538495735044872, 'E': 0.09546831152232503, 'G': 0.05798671890194771, 'H': 0.013670084187713595, 'I': 0.05907032313633964, 'W': 0.013031035536661944, 'R': 0.04623378066738907, 'D': 0.04828985024033786, 'K': 0.08852213053263316, 'N': 0.031118890833819565, 'Y': 0.028034786474396375, 'C': 0.014503625906476619}\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "cds_feature, = [feature for feature in record.features if feature.type == 'CDS']\n",
    "aa_seq, = cds_feature.qualifiers['translation']\n",
    "aa_count = Counter(aa_seq)\n",
    "aa_freq = {aa: count / len(aa_seq) for aa, count in aa_count.items()}\n",
    "print(aa_freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# URL: ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/chr11.fa.gz\n",
    "f = gzip.open(os.path.join(DIR, 'chr11.fa.gz'), 'rt')\n",
    "chr11_record, = SeqIO.parse(f, 'fasta')\n",
    "chr11_seq = chr11_record.seq\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTATCACTAAGCTCGCTTTCTTGCTGTCCAATTTCTATTAAAGGTTCCTTTGTTCCCTAAGTCCAACTACTAAACTGGGGGATATTATGAAGGGCCTTGAGCATCTGGATTCTGCCTAATAAAAAACATTTATTTTCATTGCAA\n",
      "**************************************************\n",
      "ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTATCACTAAGCTCGCTTTCTTGCTGTCCAATTTCTATTAAAGGTTCCTTTGTTCCCTAAGTCCAACTACTAAACTGGGGGATATTATGAAGGGCCTTGAGCATCTGGATTCTGCCTAATAAAAAACATTTATTTTCATTGCAA\n",
      "**************************************************\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "HBB_EXONS = [\n",
    "    (5248160, 5248301),\n",
    "    (5247807, 5248029),\n",
    "    (5246694, 5246956),\n",
    "]\n",
    "\n",
    "hbb_exon_seqs = [str(chr11_seq[(start - 1):end].reverse_complement()) for start, end in HBB_EXONS]\n",
    "hbb_recovered_seq = ''.join(hbb_exon_seqs)\n",
    "\n",
    "print(hbb_seq)\n",
    "print('*' * 50)\n",
    "print(hbb_recovered_seq)\n",
    "print('*' * 50)\n",
    "print(hbb_seq == hbb_recovered_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTATCACTAAGCTCGCTTTCTTGCTGTCCAATTTCTATTAAAGGTTCCTTTGTTCCCTAAGTCCAACTACTAAACTGGGGGATATTATGAAGGGCCTTGAGCATCTGGATTCTGCCTAATAAAAAACATTTATTTTCATTGCAA\n",
      "**************************************************\n",
      "ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTATCACTAAGCTCGCTTTCTTGCTGTCCAATTTCTATTAAAGGTTCCTTTGTTCCCTAAGTCCAACTACTAAACTGGGGGATATTATGAAGGGCCTTGAGCATCTGGATTCTGCCTAATAAAAAACATTTATTTTCATTGCAA\n",
      "**************************************************\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Repeating the same analysis, but without loading the whole chromosome into memory.\n",
    "The downside of this is that we can no longer store the sequence as a compressed file; we will have to uncompress\n",
    "chr11.fa.gz into chr11.fa.\n",
    "'''\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import Alphabet\n",
    "\n",
    "f = open(os.path.join(DIR, 'chr11.fa'), 'r')\n",
    "header_len = len(f.readline())\n",
    "line_len = len(f.readline()) - 1\n",
    "\n",
    "def convert_to_absolute_coordinate(position):\n",
    "    position_zero_index = position - 1\n",
    "    # For every 'line_len' chars, there's another \\n char to take into account.\n",
    "    return header_len + position_zero_index + (position_zero_index // line_len) \n",
    "\n",
    "def read_seq(start, end):\n",
    "\n",
    "    global f\n",
    "    \n",
    "    absolute_start = convert_to_absolute_coordinate(start)\n",
    "    absolute_length = convert_to_absolute_coordinate(end) - absolute_start + 1\n",
    "    \n",
    "    f.seek(absolute_start)\n",
    "    str_seq = f.read(absolute_length).replace('\\n', '')\n",
    "    return Seq(str_seq, Alphabet())\n",
    "    \n",
    "hbb_exon_seqs = [str(read_seq(start, end).reverse_complement()) for start, end in HBB_EXONS]\n",
    "hbb_recovered_seq = ''.join(hbb_exon_seqs)\n",
    "\n",
    "print(hbb_seq)\n",
    "print('*' * 50)\n",
    "print(hbb_recovered_seq)\n",
    "print('*' * 50)\n",
    "print(hbb_seq == hbb_recovered_seq)\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exon lengths: [142, 223, 263]\n",
      "Intron lengths: [130, 850]\n",
      "Exon GC content: 51.27\n",
      "Intron GC content: 32.96\n"
     ]
    }
   ],
   "source": [
    "hbb_introns = [(HBB_EXONS[i + 1][1] + 1, HBB_EXONS[i][0] - 1) for i in range(len(HBB_EXONS) - 1)]\n",
    "\n",
    "print('Exon lengths: ' + str([end - start + 1 for start, end in HBB_EXONS]))\n",
    "print('Intron lengths: ' + str([end - start + 1 for start, end in hbb_introns]))\n",
    "\n",
    "hbb_intron_seqs = [str(chr11_seq[(start - 1):end].reverse_complement()) for start, end in hbb_introns]\n",
    "\n",
    "print('Exon GC content: %.2f' % GC(hbb_seq))\n",
    "print('Intron GC content: %.2f' % GC(''.join(hbb_intron_seqs)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
